---
title: "Mouse Cerebellum - Reprocess progenitors"
author: "Ioannis Sarropoulos"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(viridis)
  library(Seurat)
  library(harmony)
  library(SummarizedExperiment)
  library(Matrix)
  library(irlba)
  library(RColorBrewer)
  library(ArchR)
  library(ComplexHeatmap)
})
```

This is a semi-automated script. Several decisions/entries need to be provided manually, after exploring the data!!!

```{r}
#1. Which lineage is to be processed?
cell_types <- c("Progenitor", "Progenitor_bipotent", "Progenitor_gliogenic", "Progenitor_isthmic", "Progenitor_RL", "Progenitor_VZ", "Progenitor_anteriorVZ", "astroblast_WM", "astroblast_GCL", "astro_Bergmann", "astro_parenhymal")
#2. Which marker genes to be plotted?
interesting_genes <- c("Notch1","Sox2","Fabp7", "Cyp26b1","Id3", "Id4", "Hes5", "Nes", "Slc1a3", ## General astroglia
                       "Cdon","Atoh1", "Pax6", "Reln", "Barhl1", "Wnt1", "Olig3", ## RL
                        "Dll1","Kirrel2","Ptf1a",  "Neurog2","Tfap2b", "Msx3", "Mt3","Nphs1", ## VZ
                       "Gsx1", "Pax2", "Neurod6", ## VZ interneurons
                       "Olig2","Skor2", ## VZ PCPs
                       "Robo4", "Nhlh1", "Neurog2", #GABA Nbs
                       "Slc17a6", "Meis2", "Evx1", "Barhl1",#Glut Nbs
                       "Sox10", "Lims2", ## OPCs
                       "Fgf17", "Pax5", ## Isthmic
                       "Gdf10", "Aqp4", "A2m", "Etv4", "Etv5", "Grm3","Tnc") ## Bergmann, astroglia)
#3. Minimum number of cells required in each timepoint to consider for the pseudotime
minCell <- 100
#4. Filtering of peaks - what fraction of cells in the lineage should show peak activity?
fCells <- 0.01
#5. Number of SVD components to be computed
pca_n <- 50
```

## 1. Data loading and preprocessing

Loading in data:

```{r}
peak_mat <- readRDS("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peak_matrix_SE_withInfo_robustPeaks.rds")

gsm <- readRDS("~/Mouse_Cereb/003_finalNucSet/Mouse_Cereb_finalNucSet_GeneScoreMatrix_SE.rds")

chromvar <- readRDS("~/Mouse_Cereb/007_chromvar/Mouse_Cerebellum_chromVar_cisBP_deviations_SE.rds")

cell_anno <- readRDS("~/Mouse_Cereb/004_cellTypes_broad/Mouse_Cerebellum_atac_finalCellType_annotation.rds")
cell_anno$Cell_type_color <- as.character(cell_anno$Cell_type_color)
cell_anno$Cell_type <- as.character(cell_anno$Cell_type)
cell_anno$Cell_type_color[cell_anno$Cell_type=="Progenitor_VZ"] <- "#A1BBE1"
cell_anno$Cell_type_color[cell_anno$Cell_type=="Progenitor_anteriorVZ"] <- "#795ad6"
cell_anno$Cell_type <- factor(cell_anno$Cell_type, levels = unique(cell_anno$Cell_type))
cell_anno$Cell_type_color <- factor(cell_anno$Cell_type_color, levels = unique(cell_anno$Cell_type_color))
```


We will use the standard LSI method from Cusanovich et al. 

```{r}
idf_norm = function(atac_matrix) {
  idf_counts = atac_matrix * log(1 + ncol(atac_matrix) / Matrix::rowSums(atac_matrix > 0))
  return(idf_counts)
}
```

Selecting the cells that belong to our lineage of interest:

```{r}
## Cells in lineage
cells <- cell_anno$cell[cell_anno$Cell_type %in% cell_types]
length(cells)
```

LSI and Seurat object generation

```{r}
## Subset based on cells
set.seed(1)
atac.data <- assay(peak_mat[, cells])
row.names(atac.data) <- peak_mat@elementMetadata$peak

## Requiring peaks to be active in at least a fraction of the cells (see above)
atac.data.filtered <- atac.data[Matrix::rowSums(atac.data > 0) >= round(fCells * ncol(atac.data)),]

## Dimensions of the filtered and unfiltered dataset
dim(atac.data)
dim(atac.data.filtered)
nrow(atac.data.filtered)/nrow(atac.data)

#Removing the sex chromosomes since we merge replicates from different sexes:
atac.data.filtered <- atac.data.filtered[grepl("chrX|chrY", row.names(atac.data.filtered))==F, ]
dim(atac.data.filtered)

## Then we can apply the IDF transformation
atac.data.idf <- idf_norm(atac.data.filtered)

## Computing the LSI
svd = irlba(atac.data.idf, nv = pca_n, nu = pca_n, maxit = 1000)

## Checking that we have used enough PCs (still need a few after saturation line to achieve meaningful results)
plot(svd$d/(sum(svd$d)))

## Weighting LSI based on percentage of variance explained
d_diag = matrix(0, nrow = pca_n, ncol = pca_n)
diag(d_diag) <- svd$d
lsi <- as.matrix(t(d_diag %*% t(svd$v)))

row.names(lsi) = colnames(atac.data.idf)
colnames(lsi) = paste0("LSI_",1:pca_n)

## Checking for correlation with sequencing depth
plot(apply(lsi, 2, function(x) cor(x, Matrix::colSums(atac.data.filtered))))
## The first component correlates with sequencing depth so we should remove it

## Making a Seurat object
atac.seurat <- CreateSeuratObject(atac.data, assay = "Peaks")

## LSI
atac.seurat [["lsi"]] <- CreateDimReducObject(embeddings = lsi[,2:ncol(lsi)], stdev = svd$d, key = "lsi_", assay = "Peaks")

atac.seurat <- FindNeighbors(atac.seurat, reduction = "lsi", dims = 1:(pca_n-1))
atac.seurat <- FindClusters(atac.seurat, resolution = 0.8)

atac.seurat@meta.data$clusters_original <- atac.seurat@meta.data$seurat_clusters

table(atac.seurat$seurat_clusters)

## Now running UMAP
atac.seurat <- RunUMAP(atac.seurat, dims = 1:(pca_n-1), min.dist = 0.25, n.neighbors = 30L, reduction = "lsi")

## Adding meta data
sum(cells==colnames(atac.seurat@assays$Peaks))

atac.seurat$timepoint <- peak_mat[,cells]$Timepoint
atac.seurat$sex <- peak_mat[,cells]$Sex
atac.seurat$sample_short <- peak_mat[,cells]$Sample_short

DimPlot(atac.seurat, label = T, pt.size = 0.1)
DimPlot(atac.seurat, group.by = "timepoint", pt.size = 0.1)
DimPlot(atac.seurat, group.by = "orig.ident", pt.size = 0.1)
  
##Adding gene activity and chromVar data
gsm_mat <- assay(gsm)
gsm_mat <- gsm_mat[, row.names(atac.seurat@meta.data)]
row.names(gsm_mat) <- gsm@elementMetadata$name
atac.seurat[["GeneScoreScaled"]] <- CreateAssayObject(data = gsm_mat)

chromvar_mat <- assay(chromvar, "deviations")
chromvar_mat <- chromvar_mat[, row.names(atac.seurat@meta.data)]
row.names(chromvar_mat) <- chromvar@NAMES
atac.seurat[["chromVar"]] <- CreateAssayObject(data =chromvar_mat)

rm("atac.data")
rm("atac.data.filtered")
rm("atac.data.idf")
```

Let's add our cell type annotation:

```{r}
atac.info <- atac.seurat@meta.data %>%
  as.data.frame() %>%
  rownames_to_column("cell") %>%
  left_join(select(cell_anno, cell,Cell_type, Cell_type_color)) %>%
  droplevels()

atac.seurat$Cell_type <- atac.info$Cell_type
atac.seurat$Cell_type_color <- atac.info$Cell_type_color

DimPlot(atac.seurat, group.by = "Cell_type", reduction = "umap", pt.size = 0.1) +scale_color_manual(values = levels(atac.seurat$Cell_type_color))
```

Plotting some genes

```{r}
plotMarkers <- function(gene, obj=atac.seurat) {
  c <- log10(obj@assays$GeneScoreScaled@data[gene,] + 1)
  c_new <- c
  c_new[c_new > quantile(c, 0.99)] <- quantile(c, 0.99)
  c_new[c_new < quantile(c, 0.01)] <- quantile(c, 0.01)
  ggplot(NULL, aes(x=obj@reductions$umap@cell.embeddings[,1],
                 y=obj@reductions$umap@cell.embeddings[,2],
                 col=c_new)) +
  geom_point(size=0.1, alpha=0.5) +
  scale_colour_viridis_c(na.value = "gray90", option = "B", direction = 1, name="log10 CPM") +
  xlab("UMAP1") +
  ylab("UMAP2") +
  ggtitle(paste0("GeneScore_", gene))+
  theme_classic()
}

lapply(interesting_genes, plotMarkers)
```

Making a figure:

```{r}
umap <- as.data.frame(atac.seurat@reductions$umap@cell.embeddings)

colnames(umap) <- c("UMAP1", "UMAP2")
all.df <- right_join(cell_anno, rownames_to_column(umap, "cell")) %>%
  droplevels()

all.df$Cell_type <- factor(all.df$Cell_type, levels = unique(as.character(all.df$Cell_type)))
all.df$Cell_type_color <- factor(all.df$Cell_type_color, levels = unique(as.character(all.df$Cell_type_color)))

ggplot(all.df, aes(x=UMAP1, y=UMAP2, color=Cell_type)) +
  geom_point(size=0.1, alpha=0.5) +
  scale_color_manual(values = levels(all.df$Cell_type_color)) +
  theme_classic() 
```

```{r, fig.width=5, fig.height=5}
p <- ggplot(all.df, aes(x=UMAP1, y=UMAP2, color=Cell_type)) +
  geom_point(size=0.1, alpha=0.5) +
  scale_color_manual(values = levels(all.df$Cell_type_color)) +
  theme_void() +
  theme(legend.position = "none")

p

pdf("~/Mouse_Cereb/Figures/006_astroglia/Astroglia_subtypes.pdf", width = 5, height = 5, useDingbats = F); print(p); dev.off()
```


```{r, fig.width=6, fig.height=5}
ggplot(all.df, aes(x=UMAP1, y=UMAP2, color=Timepoint)) +
  geom_point(size=0.1, alpha=0.5) +
  scale_color_brewer(palette = "Spectral") +
  theme_classic() 
```

```{r}
ggplot(all.df, aes(x=UMAP1, y=UMAP2, color=Sex)) +
  geom_point(size=0.1, alpha=0.3) +
  scale_color_manual(values=c("darkorange", "deepskyblue3", "mediumorchid")) +
  theme_classic() 
```


```{r}
p <- ggplot(all.df, aes(x=UMAP1, y=UMAP2, color=Sex)) +
  geom_point(size=0.1, alpha=0.3) +
  scale_color_manual(values=c("darkorange", "deepskyblue3", "mediumorchid")) +
  theme_void() +
  theme(legend.position = "none")

p

pdf("~/Mouse_Cereb/Figures/006_astroglia/Astroglia_sex.pdf", width = 5, height = 5, useDingbats = F); print(p); dev.off()
```


```{r, fig.width=5, fig.height=5}
p <- ggplot(all.df, aes(x=UMAP1, y=UMAP2, color=Timepoint)) +
  geom_point(size=0.1, alpha=0.5) +
  scale_color_brewer(palette = "Spectral") +
  theme_void() +
  theme(legend.position = "none")

p

pdf("~/Mouse_Cereb/Figures/006_astroglia/Astroglia_stages.pdf", width = 5, height = 5, useDingbats = F); print(p); dev.off()
```

Cell types by stage

```{r}
p <- ggplot(all.df, aes(Timepoint, fill=Cell_type)) +
  geom_bar(position = "fill") +
  scale_fill_manual(values = levels(all.df$Cell_type_color)) +
  ylab("Fraction") +
  xlab("Developmental stage") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

p

pdf("~/Mouse_Cereb/Figures/006_astroglia/Astroglia_CellTypes_byStage.pdf", width = 6, height = 3.5, useDingbats = F); print(p); dev.off()
```

Fraction of progenitors per timepoint:

```{r}
prog_f <- group_by(cell_anno, Timepoint) %>%
  summarise(f_prog=sum(cell %in% all.df$cell)/n())

prog_f$t <- as.numeric(as.factor(prog_f$Timepoint))

p <- ggplot(prog_f, aes(x=t, y=f_prog)) +
  geom_point(color="#BB4D90") +
  scale_x_continuous(breaks = unique(prog_f$t), labels = unique(prog_f$Timepoint)) +
  geom_line(color="#BB4D90") +
  ylab("Astroglia fraction") +
  xlab("Developmental stage") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

p

pdf("~/Mouse_Cereb/Figures/006_astroglia/Astroglia_fraction_byStage.pdf", width = 4, height = 2, useDingbats = F); print(p); dev.off()

```


```{r}
saveRDS(atac.seurat, "~/Mouse_Cereb/014_Astroglia/Mouse_progenitors_Seurat.rds")
```

Also identifying marker genes per subtype (for this we can switch back to ArchR).

```{r}
proj <- loadArchRProject("~/Mouse_Cereb/proj5_final/")

proj_astro <- subsetCells(proj, cellNames = row.names(atac.seurat@meta.data[!(atac.seurat@meta.data$timepoint %in% c("P14", "P63")),]))

sum(proj_astro$cellNames == row.names(atac.seurat@meta.data[!(atac.seurat@meta.data$timepoint %in% c("P14", "P63")),]))

proj_astro$comp <- paste(as.character(atac.seurat$Cell_type[!(atac.seurat@meta.data$timepoint %in% c("P14", "P63"))]), as.character(atac.seurat$timepoint[!(atac.seurat@meta.data$timepoint %in% c("P14", "P63"))]), sep = "-")

major_comp <- names(which(table(proj_astro$comp) >= 50))

markersGS_cell_types <-getMarkerFeatures(proj_astro,
                  groupBy = "comp",
                  useMatrix = "GeneScoreMatrix", 
                  useGroups = major_comp, 
                  bgdGroups = major_comp,
                  threads = 20,
                  scaleTo = 10^4)

markersGS_cell_types_se <- getMarkers(markersGS_cell_types, cutOff = "FDR < 0.01 & Log2FC > 1")
```

```{r, fig.width=8, fig.height=11}
mark_stats <- as.matrix(assay(markersGS_cell_types))
row.names(mark_stats) <-as.character(markersGS_cell_types@elementMetadata$name)

ct <- major_comp

top_markers <- na.omit(unique(unlist(sapply(ct, function(n) {
  x <- markersGS_cell_types_se[[n]]
  x.df <- data.frame(name=as.character(x$name),
                     FDR=x$FDR,
                     Log2FC=x$Log2FC, stringsAsFactors = F) %>%
    mutate(score=Log2FC * (-log10(FDR))) %>%
    filter(grepl("Rik$|^Gm[0-9]*", name)==F) %>%
    arrange(desc(score))
    return(as.character(x.df$name[1:min(3, nrow(x))]))
}
))))

sort(top_markers)

mark_stats.plot <- mark_stats[top_markers,ct]

## Cap to 2 /0
mark_stats.plot[mark_stats.plot > 2] <- 2
mark_stats.plot[mark_stats.plot < 0] <- 0

cell_info <- dplyr::select(cell_anno, Cell_type, Cell_type_color, Cell_type_broad, Cell_type_broad_color) %>%
  unique()

cell_type <- sapply(colnames(mark_stats.plot), function(x) str_split(x, "-", simplify = T)[1])

cell_type

cell_type_cols <- as.character(sapply(unique(cell_type), function(x) cell_info$Cell_type_color[cell_info$Cell_type==x]))
names(cell_type_cols) <- unique(cell_type)

timepoint <- sapply(colnames(mark_stats.plot), function(x) str_split(x, "-", simplify = T)[2])

timepoint

timepoint_cols <- RColorBrewer::brewer.pal(length(unique(timepoint)), "Spectral")
names(timepoint_cols) <- sort(unique(timepoint))

ha = HeatmapAnnotation(timepoint=timepoint,cell_type = cell_type, col=list(cell_type=cell_type_cols, timepoint=timepoint_cols),annotation_name_side = "right", annotation_legend_param = list(direction = "vertical",ncol = 1))

draw(Heatmap(matrix = mark_stats.plot,
        cluster_rows = T,
        cluster_columns = T,
        clustering_distance_rows = "pearson", 
        clustering_distance_columns = "pearson",
        clustering_method_columns = "ward.D2", 
        clustering_method_rows = "ward.D2",
        show_row_names = T,
        show_column_names = T,
        col=viridis(100,option = "B", direction = 1),
        border = TRUE,
        top_annotation = ha,
        heatmap_legend_param = list(direction = "vertical", title = "Z-score GS")), heatmap_legend_side = "right")
```

```{r}
lapply(c("Pax5", "Gsx1", "Ptf1a", "Cdon", "Slc1a3", "Gdf10", "Ndnf", "Robo1", "Wnt8b"), plotMarkers)
```

```{r}
if (dir.exists("~/Mouse_Cereb/014_Astroglia/Marker_genes_gs")==F) {
  dir.create("~/Mouse_Cereb/014_Astroglia/Marker_genes_gs")
}

saveRDS(markersGS_cell_types, "~/Mouse_Cereb/014_Astroglia/Marker_genes_gs/markers_byType_andStage.rds")
saveRDS(markersGS_cell_types_se, "~/Mouse_Cereb/014_Astroglia/Marker_genes_gs/markers_byType_andStage_se_FDR0.01_LogFC_1.rds")
```


```{r}
sessionInfo()
```

