---
title: "Mouse Cerebellum - Reprocess by developmental stage"
author: "Ioannis Sarropoulos"
date: '`r format(Sys.time(), "%d %B, %Y")`'
params:
  timepoint:
    value: x
output: html_document
---

Setting the root directory to the location where our arrow files are stored.

````{r "setup", include=FALSE}
require("knitr")
opts_knit$set(root.dir = "~/Mouse_Cereb/")
```

```{r}
suppressPackageStartupMessages({
  library(ArchR)
  library(tidyverse)
  library(RColorBrewer)
  library(ComplexHeatmap)
  library(viridis)
})
```

Stage to be reprocessed:

```{r}
timepoint <- params$timepoint
print(paste0("Reprocessing stage: ", timepoint))
```

Initialising ArchR

```{r}
addArchRThreads(threads = 10) 

## Genome and annotation
addArchRGenome("mm10")
```

We can directly use our previously generated arrow files:

```{r}
ArrowFiles <- list.files(".",".arrow$")
## Subsetting for our developmental stage
ArrowFiles <- ArrowFiles[grepl(timepoint, ArrowFiles)]
```

We can directly initialise the project.

```{r}
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = paste0("010_byStage/proj_", timepoint),
  copyArrows = TRUE #This is recommened so that you maintain an unaltered copy for later usage.
)

## Also adding more metadata
proj$cellNames_short <- sapply(getCellNames(proj), function(x) {
  paste(str_extract(x, "sa[0-9]*"), str_split(x, pattern = fixed("#"), simplify = T)[2], sep="_")
})

proj$Sample_short <- sapply(proj$Sample, function(x) {
  x <- paste(str_split(x, pattern = fixed("_"), simplify = T)[c(4, 1, 5)], collapse="_")
  x <- gsub("P4", "P04", x)
  x <- gsub("P7", "P07", x)
})

proj$Timepoint <- sapply(proj$Sample_short, function(x) str_split(x, pattern = fixed("_"), simplify = T)[1])
proj$Lib <- sapply(proj$Sample_short, function(x) str_split(x, pattern = fixed("_"), simplify = T)[2])
proj$Sex <- sapply(proj$Sample_short, function(x) str_split(x, pattern = fixed("_"), simplify = T)[3])
```

Now we should subset for the cells that passed QC:

```{r}
barcode_stats <- readRDS("001_barcode_qc/Mouse_Cerebellum_barcode_stats.rds")
proj <- subsetCells(proj,
                    cellNames = intersect(
                      proj$cellNames,
                      barcode_stats$barcode[barcode_stats$PassQC_all==T]
                      )
                    )
```


Adding iterative LSI. Here we are analysing a single developmental stage. Thus, we can reduce the number of dimensions and features to be used.

```{r}
proj <- addIterativeLSI(ArchRProj = proj,
                        useMatrix = "TileMatrix",
                        name = paste0("IterativeLSI_", timepoint),
                        iterations=5,
                        clusterParams = list(
                          resolution = c(0.1, 0.2, 0.4, 0.8), 
                          sampleCells = 10000, 
                          n.start = 10
                        ), 
                        varFeatures = 50000,
                        dimsToUse = 1:50,
                        totalFeatures = 200000,
                        seed = 1,
                        LSIMethod = 1,
                        scaleDims = FALSE,
                        corCutOff = 0.75,
                        excludeChr = c("chrX", "chrY", "chrMT"),
                        binarize = T,
                        force = T)
```

For postnatal samples, we have observed mild batch effects that only become apparent when considering individual timepoints. This is likely due to the reduced cellular complexity in P14 and P63 samples (mostly GCs). However, for consistency, let's include a harmony correction of LSI in all timepoints:

```{r}
proj <- addHarmony(proj,
                   reducedDims = paste0("IterativeLSI_", timepoint),
                   name =paste0("HarmonyLSI_", timepoint))
```

Adding clusters. We can also reduce the resolution to 1.0:

```{r}
proj <- addClusters(input = proj,
                    name = paste0("Clusters_within_", timepoint),
                    reducedDims = paste0("HarmonyLSI_", timepoint),
                    method = "Seurat",
                    force = T,
                    resolution=1.0,
                    corCutOff = 0.75,
                    scaleDims = FALSE,
                    seed = 1)
```

And a UMAP. Here we are looking for smaller scale substructure so we can reduce the number of neighbors:

```{r}
proj <- addUMAP(ArchRProj = proj,
                name = paste0("UMAP_within_", timepoint),
                reducedDims = paste0("HarmonyLSI_", timepoint),
                minDist = 0.15,
                metric = "cosine",
                nNeighbors = 15,
                force = T,
                seed = 1,
                scaleDims = F,
                corCutOff = 0.75)
```

Plotting

```{r, fig.width=8, fig.height=8}
p1 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Sample_short", embedding = paste0("UMAP_within_", timepoint), rastr = T, labelMeans = F)
p2 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = paste0("Clusters_within_", timepoint), embedding = paste0("UMAP_within_", timepoint), rastr = T)

ggAlignPlots(p1, p2, type = "h")
```

```{r, fig.width=5, fig.height=5}
interesting_genes <- c("Notch1","Sox2","Fabp7","Cdon","Atoh1","Slc17a6", "Kirrel2","Gad2", "Meis1","Meis2", "Neurod1","Neurod2","Neurod6","Nr2f1","Lmo3","Lmx1a", "Lmx1b", "Pax5", "Pax8","Pax2", "Top2a", "Ptf1a","Rora", "Rorb", "Tfap2b","Isl1","Isl2","Lef1","Pou6f2","Tlx3", "Evx1", "Evx2", "Sox14", "Skor2", "Lhx5","Aqp4", "Olig2", "Cbln3", "Fgf3", "Grin2b","Nhlh1", "Gdf10","Map3k1", "Gli2", "Esrrg", "Sorcs3", "Cdh22","Foxp1", "Cdh9", "March11", "Pvalb", "Lhx2","Gabra6", "Slc1a3", "Inpp5d", "Gsx1")

interesting_genes <- interesting_genes[interesting_genes %in% proj@geneAnnotation$genes$symbol]

p <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = interesting_genes, 
    embedding = paste0("UMAP_within_", timepoint),
    quantCut = c(0.01, 0.95),
    imputeWeights = NULL
)

p
```

Adding our peaks and peak matrix in case we need them later:

```{r}
peaks <- readRDS("006_peaks/Mouse_Cerebellum_reproducible_peaks.rds")

length(peaks)

proj <- addPeakSet(proj, 
                   peaks,
                   force = T)

proj <- addPeakMatrix(proj,
                      ceiling = 5,
                      binarize = F,
                      force = T)
```

Finding marker genes between the clusters

```{r}
markersGS_clusters <-getMarkerFeatures(proj,
                  groupBy = paste0("Clusters_within_", timepoint),
                  useGroups = names(which(table(getCellColData(proj, paste0("Clusters_within_", timepoint), drop = T)) > 20)),
                  useMatrix = "GeneScoreMatrix",
                  threads = 10,
                  scaleTo = 10^4)
markersGS <- getMarkers(markersGS_clusters, cutOff = "FDR < 0.01 & Log2FC > 1")
```

```{r, fig.height=12, fig.width=7}
mark_stats <- as.matrix(assay(markersGS_clusters))
row.names(mark_stats) <-as.character(markersGS_clusters@elementMetadata$name)

top_markers <- na.omit(unique(as.character(sapply(markersGS, function(x) as.character(x$name[grepl("^Gm[0-9]|Rik$|^Mir[0-9]|^Olfr", x$name)==F][1:5])))))

mark_stats.plot <- mark_stats[top_markers,]

## Cap to 2 /-1
mark_stats.plot[mark_stats.plot > 2] <- 2
mark_stats.plot[mark_stats.plot < (-1)] <- (-1)

draw(Heatmap(matrix = mark_stats.plot,
        cluster_rows = T,
        cluster_columns = T,
        clustering_method_columns = "ward.D2",
        clustering_method_rows = "ward.D2",
        show_row_names = T,
        show_column_names = T,
        col=viridis(100,option = "B"),
        border = TRUE,
        heatmap_legend_param = list(direction = "vertical", title = "Z-score GS")), heatmap_legend_side = "right")
```

We can save the project

```{r}
saveArchRProject(proj)
```

And the marker gene object

```{r}
saveRDS(markersGS_clusters, paste0("010_byStage/atac_seurat/Mouse_Cerebellum_", timepoint, "_markerGenes.rds"))
```


Now we can also create a Seurat object to be used for the integration with the RNA-seq data.
This will allow us more flexibility and to perform the integration by timepoint rather than in the merged dataset.

We only need the gene score matrix, embeddings and metadata:

```{r}
gsm <- getMatrixFromProject(proj,
                            useMatrix = "GeneScoreMatrix")

gsm_mat <- assay(gsm)
gsm_mat <- gsm_mat[, proj$cellNames]
colnames(gsm_mat) <- proj$cellNames_short 
row.names(gsm_mat) <- gsm@elementMetadata$name

## Need to swap to Ensembl IDs to facilitate integration with our other datasets
gene.symbols <- read.table("~/Data/scATAC_pipeline_v3/resources/mouse_coding_ens94_gene_symbols.txt", header = T, sep = "\t", stringsAsFactors = F)

atac.genes <- data.frame(symbol = row.names(gsm_mat), stringsAsFactors = F) %>%
  left_join(gene.symbols) %>%
  group_by(symbol) %>%
  summarise(ensembl=ensembl[1]) %>%
  ungroup()

sum(is.na(atac.genes$ensembl))

## IF Ensembl is NA, keep using the symbol (these genes will anyway be discarded for integration)

atac.genes$new_name <- atac.genes$ensembl
atac.genes$new_name[is.na(atac.genes$ensembl)] <- atac.genes$symbol[is.na(atac.genes$ensembl)]

length(unique(atac.genes$new_name))==length(atac.genes$new_name)
length(unique(atac.genes$new_name))==length(row.names(gsm_mat))
sum(atac.genes$symbol==row.names(gsm_mat))

gsm_mat <- gsm_mat[atac.genes$symbol, ]
sum(atac.genes$symbol==row.names(gsm_mat))
row.names(gsm_mat) <- atac.genes$new_name

## Extracting (harmony corrected) LSI
lsi <- getReducedDims(proj, 
                      paste0("HarmonyLSI_", timepoint), 
                      corCutOff = 0.75,
                      scaleDims = FALSE)

row.names(lsi) <- proj$cellNames_short 

metad <- as.data.frame(getCellColData(proj, 
                        c(paste0("Clusters_within_", timepoint), "Timepoint", "Lib", "Sex")))

row.names(metad) <- proj$cellNames_short 

umap <- as.matrix(getEmbedding(proj, paste0("UMAP_within_", timepoint)))
row.names(umap) <- proj$cellNames_short 
colnames(umap) <- c("UMAP1", "UMAP2")

atac.seurat <- CreateSeuratObject(counts = gsm_mat, meta.data = metad, assay = "GeneScoreScaled")
atac.seurat[["lsi"]] <- CreateDimReducObject(embeddings = lsi, key = "lsi_", assay = "GeneScoreScaled")
atac.seurat[["umap"]] <- CreateDimReducObject(embeddings = umap, key = "umap_", assay = "GeneScoreScaled")

saveRDS(atac.seurat, paste0("010_byStage/atac_seurat/Mouse_Cerebellum_", timepoint, "_atac_seurat.rds"))
```

```{r}
sessionInfo()
```
