---
title: "Mouse Cerebellum 010 - Peak annotation"
author: "Ioannis Sarropoulos"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(ArchR)
  library(tidyverse)
  library(RColorBrewer)
  library(viridis)
  library(ComplexHeatmap)
  library(data.table)
})
```

Here we will annotate our peaks based on their genomic and conservation features.

```{r}
setwd("~/Mouse_Cereb")

union.peaks <- readRDS("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peaks.rds")

length(union.peaks)
```

Let's export a bed file that we can use for our annotation:

```{r}
peak_names <- paste(seqnames(union.peaks), start(ranges(union.peaks)), end(ranges(union.peaks)), sep = "_")
peak.bed <- data.frame(chr=seqnames(union.peaks),
                       start=start(ranges(union.peaks)),
                       end=end(ranges(union.peaks)),
                       peak=peak_names, stringsAsFactors = F)

write.table(peak.bed, "006_peaks/Mouse_Cerebellum_reproducible_peaks.bed",col.names = F, row.names = F, sep = "\t", quote = F)
```

```{r}
peaks <- as.data.frame(union.peaks@elementMetadata)
peaks$peak=peak_names

peaks <- left_join(peak.bed, peaks)
```

## Genomic context

We already have genomic annotations from ArchR based on UCSC annotations.

We can supplement this by:
1) Including our lncRNAs from our recent Nature paper
2) Include the biotype of the associated genes

#### Disclaimer: ArchR uses "distal" for intergenic elements. In the manuscript we collectively refer to intergenic and intronic elements as distal elements, as opposed to promoters and exonic peaks.

LncRNA intersection 

```{bash}
bedtools intersect -a 006_peaks/Mouse_Cerebellum_reproducible_peaks.bed -b 006_peaks/001_peak_annotation/genes/mouse.lncRNA.exons.bed -wa -wb > 006_peaks/001_peak_annotation/genes/Mouse_Cerebellum_reproducible_peaks_overlap_XLOC_exons.txt

bedtools intersect -a 006_peaks/Mouse_Cerebellum_reproducible_peaks.bed -b 006_peaks/001_peak_annotation/genes/mouse.lncRNA.genes.bed -wa -wb > 006_peaks/001_peak_annotation/genes/Mouse_Cerebellum_reproducible_peaks_overlap_XLOC_genes.txt

bedtools intersect -a 006_peaks/Mouse_Cerebellum_reproducible_peaks.bed -b 006_peaks/001_peak_annotation/genes/mouse.lncRNA.promoters_up2Kbdown100bp.bed -wa -wb > 006_peaks/001_peak_annotation/genes/Mouse_Cerebellum_reproducible_peaks_overlap_XLOC_promoters.txt
```

Read overlaps

```{r}
exon.overlaps <- fread("006_peaks/001_peak_annotation/genes/Mouse_Cerebellum_reproducible_peaks_overlap_XLOC_exons.txt")
gene.overlaps <- fread("006_peaks/001_peak_annotation/genes/Mouse_Cerebellum_reproducible_peaks_overlap_XLOC_genes.txt")
promoter.overlaps <- fread("006_peaks/001_peak_annotation/genes/Mouse_Cerebellum_reproducible_peaks_overlap_XLOC_promoters.txt")
```

```{r}
length(unique(exon.overlaps$V4))
length(unique(gene.overlaps$V4))
length(unique(promoter.overlaps$V4))
```

We will only update the annotation for peaks previously assigned as distal or intronic.

```{r}
prom_i <- which(peaks$peak %in% promoter.overlaps$V4 & peaks$peakType %in% c("Distal", "Intronic"))
exon_i <- which(peaks$peak %in% exon.overlaps$V4 & peaks$peakType %in% c("Distal", "Intronic"))
exon_i <- exon_i[!(exon_i %in% prom_i)]
intron_i <- which(peaks$peak %in% gene.overlaps$V4 & peaks$peakType %in% c("Distal"))
intron_i <- intron_i[!(intron_i %in% c(exon_i, prom_i))]

length(prom_i)
length(exon_i)
length(intron_i)

peaks$genomic_class_broad <- peaks$peakType
peaks$genomic_class_broad[prom_i] <- "Promoter"
peaks$genomic_class_broad[exon_i] <- "Exonic"
peaks$genomic_class_broad[intron_i] <- "Intronic"


table(peaks$peakType)
table(peaks$genomic_class_broad)

peaks$associated_gene <- peaks$nearestGene
peaks$associated_gene[prom_i] <- sapply(peaks$peak[prom_i], function(x) promoter.overlaps$V8[promoter.overlaps$V4==x][1])
peaks$associated_gene[exon_i] <- sapply(peaks$peak[exon_i], function(x) exon.overlaps$V8[exon.overlaps$V4==x][1])
peaks$associated_gene[intron_i] <- sapply(peaks$peak[intron_i], function(x) gene.overlaps$V8[gene.overlaps$V4==x][1])

sum(grepl("XLOC", peaks$associated_gene ))

peaks$closest_TSS_dist <- sapply(1:nrow(peaks), function(i) {
  ifelse(peaks$genomic_class_broad[i] == "Promoter" & peaks$peakType[i]!="Promoter", 
         yes = 0,
         no = peaks$distToTSS[i])
})

peaks$closest_TSS <- sapply(1:nrow(peaks), function(i) {
  ifelse(peaks$genomic_class_broad[i] == "Promoter" & peaks$peakType[i]!="Promoter", 
         yes = peaks$associated_gene[i],
         no = peaks$nearestTSS[i])
})

peaks <- select(peaks, -peakType, -nearestGene, -distToTSS, -nearestTSS)
```

We can also incorporate the biotype annotation.

```{r}
biomart.biotypes <- read_tsv("006_peaks/001_peak_annotation/genes/mm10_gene_biotypes_bioMart_Ens100_200508.txt", col_names = c("Ensembl", "associated_gene_biotype", "associated_gene"), skip = 1)

length(unique(biomart.biotypes$associated_gene))
nrow(biomart.biotypes)

## Very few genes have two Ensembl IDs and thus biotypes (mostly lncRNAs) -> for simplicity, let's keep the first only:
biomart.biotypes <- group_by(biomart.biotypes, associated_gene) %>%
  summarise(Ensembl= Ensembl[1], associated_gene_biotype= associated_gene_biotype[1]) %>%
  ungroup()

## Collaping temrs into more useful categories
sort(table(biomart.biotypes$associated_gene_biotype), decreasing = T)
biomart.biotypes$associated_gene_biotype[biomart.biotypes$associated_gene_biotype %in% c("lincRNA", "bidirectional_promoter_lncRNA", "lncRNA", "antisense")] <- "lncRNA"
biomart.biotypes$associated_gene_biotype[biomart.biotypes$associated_gene_biotype %in% c("miRNA", "snoRNA", "snRNA")] <- "smallRNA"
biomart.biotypes$associated_gene_biotype[!(biomart.biotypes$associated_gene_biotype %in% c("protein_coding", "lncRNA", "smallRNA"))] <- "other"

table(biomart.biotypes$associated_gene_biotype)

length(unique(biomart.biotypes$associated_gene))

peaks <- left_join(peaks, select(biomart.biotypes, associated_gene, associated_gene_biotype))
table(peaks$associated_gene_biotype)
peaks$associated_gene_biotype[peaks$genomic_class_broad=="Distal"] <- NA
table(peaks$associated_gene_biotype)
peaks$associated_gene_biotype[grepl("XLOC", peaks$associated_gene)] <- "lncRNA"

peaks$genomic_class <- paste(peaks$associated_gene_biotype, peaks$genomic_class_broad, sep =  "-")
peaks$genomic_class <- gsub("^NA-", "", peaks$genomic_class)
table(peaks$genomic_class)
peaks$genomic_class[peaks$genomic_class=="Promoter"] <- "other-Promoter"
peaks$genomic_class[peaks$genomic_class=="Exonic"] <- "other-Exonic"
peaks$genomic_class[peaks$genomic_class=="Intronic"] <- "other-Intronic"

table(peaks$genomic_class)
```

## Distance to adjacent peaks

This is important since we have required a fixed width of 500 bp for our peaks. This means that if we have wider CREs, they might be split over multiple peaks. 

We can incorporate such peaks into a functional unit, provided they are adjacent and co-accessible.

```{r}
peaks$distance_to_next_peak <- c(sapply(1:(nrow(peaks)-1), function(i) {
  if (peaks$chr[i] == peaks$chr[i+1]) {
    return(peaks$start[i+1] - peaks$end[i])
  } else {
    return(NA)
  }
}), NA)

p <- ggplot(peaks, aes(log10(distance_to_next_peak))) +
  geom_histogram(bins = 100, color="deepskyblue3", fill="deepskyblue3", alpha=0.3) +
  geom_vline(xintercept = 3, color="red", lty="dashed") +
  xlab("Distance to downstream peak (log10 bp)") +
  theme_classic()

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_distance_to_downstream_peak.pdf", width = 4, height = 3, useDingbats = F); print(p); dev.off()

summary(peaks$distance_to_next_peak)
sum(na.omit(peaks$distance_to_next_peak) < 1000)/length(na.omit(peaks$distance_to_next_peak))
```

We see that we have a mixed distribution with many (~55%) of the peaks being very close to each other (up to 1,000 bp), whereas some peaks are very far apart (up to 700Kb)!

We should consider uniting peaks that are very close to each other into a functional unit.

We will use the results from the coaccessibility analysis for this.

```{r}
cicero_conns <- readRDS("~/Mouse_Cereb/008_cicero/Mouse_Cerebellum_cicero_conns_all.rds")
sum(peaks$peak == paste(seqnames(cicero_conns@metadata$peakSet),
                        start(ranges(cicero_conns@metadata$peakSet)), 
                        end(ranges(cicero_conns@metadata$peakSet)), sep = "_"))

peaks$idx_all <- 1:nrow(peaks)
head(cicero_conns)

corr_to_next <- data.frame(queryHits=1:(nrow(peaks)-1), subjectHits=(1:(nrow(peaks)-1)+1), stringsAsFactors = F) %>%
  left_join(as.data.frame(cicero_conns))

peaks$corr_to_next_peak <- c(corr_to_next$correlation,NA)
peaks$corr_to_next_peak[is.na(peaks$distance_to_next_peak)] <- NA

set.seed(1)
p <- ggplot(filter(peaks, peak %in% sample(peaks$peak, 2e4)), aes(x=log10(distance_to_next_peak),
                                                             y=corr_to_next_peak, color=as.factor(distance_to_next_peak <=1000 & corr_to_next_peak >= 0.4))) +
  geom_point(alpha=0.1) +
  scale_color_manual(values = c("gray40", "deepskyblue4"), name="in CRE unit", labels=c("no","yes")) +
  geom_vline(xintercept = log10(1000), color="red", lty="dashed") +
  geom_hline(yintercept = 0.4, color="red", lty="dashed") +
  xlab("Distance to downstream peak (log10 bp)") +
  ylab("Correlation to downstream peak") +
  theme_bw()

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_Coaccessibility_vs_distance_to_downstream_peak.pdf", width = 6, height = 5, useDingbats = F); print(p); dev.off()

sum(peaks$distance_to_next_peak[is.na(peaks$distance_to_next_peak)==F] <= 1000 & peaks$corr_to_next_peak[is.na(peaks$distance_to_next_peak)==F] >= 0.4)
```

Constructing these broad CRE units (chaining across peaks)

```{r}
chain_query_idx <- filter(peaks, distance_to_next_peak <=1000, corr_to_next_peak >=0.4)$idx_all
chain_subj_idx <- chain_query_idx + 1

i <- 1
new_chain <- c()
chains <- list()
while(i < (length(chain_query_idx))){
  new_chain <- unique(c(new_chain, c(chain_query_idx[i], chain_subj_idx[i])))
  if(chain_subj_idx[i] == (chain_query_idx[i + 1])) {
    i <- i + 1
  } else {
    chains <- c(chains, list(new_chain))
    new_chain <- c()
    i <- i + 1
  }
}

length(chains)

summary(sapply(chains, length))

max(table(unlist(chains)))

names(chains) <- paste0("broadCRE", 1:length(chains))

chain_df <- data.frame(idx_all=unlist(chains), multiPeakCRE=rep(names(chains), sapply(chains, length)),stringsAsFactors = F)

chain_df[1:20,]

## Now we can incorporate this into our peaks summary
peaks <- left_join(peaks, chain_df)
table(is.na(peaks$multiPeakCRE))

peaks$multiPeakCRE[is.na(peaks$corr_to_next_peak)] <- NA
table(is.na(peaks$multiPeakCRE))
```


We can also calculate the coordinates and width of these broad CREs

```{r}
peaks <- group_by(peaks, multiPeakCRE) %>%
  mutate(multiPeakCRE_start=min(start), 
         multiPeakCRE_end=max(end),
         multiPeakCRE_width=max(end)-min(start),
         multiPeakCRE_nPeaks=n_distinct(peak)) %>%
  ungroup()

peaks$multiPeakCRE_start[is.na(peaks$multiPeakCRE)] <- NA
peaks$multiPeakCRE_end[is.na(peaks$multiPeakCRE)] <- NA
peaks$multiPeakCRE_width[is.na(peaks$multiPeakCRE)] <- NA
peaks$multiPeakCRE_nPeaks[is.na(peaks$multiPeakCRE)] <- NA

summary(peaks$multiPeakCRE_width)
summary(peaks$multiPeakCRE_nPeaks)

```

## Sequence constraint

We can use phastCons scores to estimate conservation of each peak.
(http://hgdownload.cse.ucsc.edu/goldenpath/mm10/phastCons60way/mm10.60way.phastCons.bw)

Since we don't necessarily expect the entire peak to be under constraint, we are using a simple sliding window approach (creating 100-bp windows sliding the start position every 10 bp).

Our peaks have a fixed width so we are fair in our testing (same number of windows per peak).

```{r}
sliding_window <- function(start, end, step=10, width=100) {
  start_pos <- seq(from = start, to = end - width, by = step)
  return(start_pos)
}

start_pos <- parallel::mclapply(1:nrow(peaks), function(i) {
  pos_i <- sliding_window(peaks$start[i], peaks$end[i])
  names(pos_i) <- paste(peaks$peak[i], 1:length(pos_i), sep = "_")
  return(drop(pos_i))
}, mc.cores = 5)

start_pos <- unlist(start_pos)

peak_sliding_windows <- data.frame(window_name=names(start_pos), start=start_pos, end=start_pos + 100, stringsAsFactors = F) %>%
  mutate(peak=gsub("_[0-9]*$", "", window_name)) %>%
  left_join(select(peaks, peak, chr))

peak_sliding_windows <- peak_sliding_windows[, c("chr", "start", "end", "window_name")]

write_tsv(peak_sliding_windows, "006_peaks/001_peak_annotation/phastCons/Mouse_Cerebellum_peaks_final_100bp_windows_10bp_step.bed", col_names = F)
```

Calculating average phastCons score for each window:

```{bash}
~/Software/utils/bigWigAverageOverBed 006_peaks/001_peak_annotation/phastCons/mm10.60way.phastCons.bw 006_peaks/001_peak_annotation/phastCons/Mouse_Cerebellum_peaks_final_100bp_windows_10bp_step.bed 006_peaks/001_peak_annotation/phastCons/Mouse_Cerebellum_peaks_100bp_windows_phastCons_60way.txt -minMax
```

Reading in the results:

```{r}
phastcons_windows <- read_tsv("006_peaks/001_peak_annotation/phastCons/Mouse_Cerebellum_peaks_100bp_windows_phastCons_60way.txt", col_names = F)

## First let's check the width of the windows
summary(phastcons_windows$X2)

## The column of interest is X6 (mean0, i.e.counting non-covered bases as 0)
phastcons_windows_sum <- mutate(phastcons_windows, peak=gsub("_[0-9]*$", "", X1)) %>%
  group_by(peak) %>%
  summarise(meanPhastCons_100bp=max(X6), mostCons100bp_window=X1[which.max(X6)])

ggplot(phastcons_windows_sum, aes(meanPhastCons_100bp))+
  geom_histogram()

## Getting the info about the most conserved region:
phastcons_windows_sum <- left_join(phastcons_windows_sum, rename(peak_sliding_windows,mostCons100bp_window=window_name))

phastcons_windows_sum$mostCons100bp <- paste(phastcons_windows_sum$chr, phastcons_windows_sum$start, phastcons_windows_sum$end, sep = "_")

peaks <- left_join(peaks, select(phastcons_windows_sum, peak, meanPhastCons_100bp, mostCons100bp))
```

Conservation by genomic class

```{r}
p <- ggplot(peaks, aes(x=genomic_class, y=meanPhastCons_100bp, fill=genomic_class_broad)) +
  geom_violin(alpha=0.5) +
  geom_boxplot(notch = T, width=0.2, outlier.alpha = 0) +
  theme_classic() +
  xlab("") +
  ylab("PhastCons (vertebrates)") +
  scale_fill_manual(values = c("darkorange2", "deepskyblue3", "darkolivegreen3","darkorchid"), name="Broad\nGenomic\nClass") +
  theme(axis.text.x = element_text(angle=45, hjust = 0.95))

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_phastCons_vertebrates_byGenomicClass.pdf", width = 8, height = 5, useDingbats = F); print(p); dev.off()
```

We see that intronic CREs, lncRNA promoters and exons and distal peaks show very similar distributions.

On the other hand, protein-coding, small RNA and other promoters are overall more conserved and should be treated differently.

Finally, exonic CREs show very high conservation due to the overlap with CDS regions that is probably irrelevant with any regulatory roles. Thus, such regions should be excluded from all conservation analyses!

Let's repeat this with phastCons in glires:

```{bash}
~/Software/utils/bigWigAverageOverBed 006_peaks/001_peak_annotation/phastCons/mm10.60way.phastCons60wayGlire.bw 006_peaks/001_peak_annotation/phastCons/Mouse_Cerebellum_peaks_final_100bp_windows_10bp_step.bed 006_peaks/001_peak_annotation/phastCons/Mouse_Cerebellum_peaks_100bp_windows_phastCons_60way_Glires.txt -minMax
```

```{r}
phastcons_windows_glires <- read_tsv("006_peaks/001_peak_annotation/phastCons/Mouse_Cerebellum_peaks_100bp_windows_phastCons_60way_Glires.txt", col_names = F)

## First let's check the width of the windows
summary(phastcons_windows_glires$X2)

## The column of interest is X6 (mean0, i.e.counting non-covered bases as 0)
phastcons_windows_glires_sum <- mutate(phastcons_windows_glires, peak=gsub("_[0-9]*$", "", X1)) %>%
  group_by(peak) %>%
  summarise(meanPhastCons_Glires_100bp=max(X6), mostCons100bp_window=X1[which.max(X6)])

ggplot(phastcons_windows_glires_sum, aes(meanPhastCons_Glires_100bp))+
  geom_histogram()

## Getting the info about the most conserved region:
phastcons_windows_glires_sum <- left_join(phastcons_windows_glires_sum, rename(peak_sliding_windows,mostCons100bp_window=window_name))

phastcons_windows_glires_sum$mostCons100bp_Glires <- paste(phastcons_windows_glires_sum$chr, phastcons_windows_glires_sum$start, phastcons_windows_glires_sum$end, sep = "_")

peaks <- left_join(peaks, select(phastcons_windows_glires_sum, peak, meanPhastCons_Glires_100bp, mostCons100bp_Glires))
```

```{r}
p <- ggplot(peaks, aes(x=genomic_class, y=meanPhastCons_Glires_100bp, fill=genomic_class_broad)) +
  geom_violin(alpha=0.5) +
  geom_boxplot(notch = T, width=0.2, outlier.alpha = 0) +
  theme_classic() +
  xlab("") +
  ylab("PhastCons (glires)") +
  scale_fill_manual(values = c("darkorange2", "deepskyblue3", "darkolivegreen3","darkorchid"), name="Broad\nGenomic\nClass") +
  theme(axis.text.x = element_text(angle=45, hjust = 0.95))

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_phastCons_glires_byGenomicClass.pdf", width = 8, height = 5, useDingbats = F); print(p); dev.off()
```

Same conclusions regarding genomic classes.

```{r}
cor(peaks$meanPhastCons_100bp, peaks$meanPhastCons_Glires_100bp)
```

## Age of DNA region

```{bash, eval=FALSE}
sh 006_peaks/001_peak_annotation/liftover_vertebrates/liftover_vertebrates.sh
```

Reading in liftover files

```{r}
mouse_peaks <- read_tsv("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peaks.bed", col_names = c("chr", "start", "end", "peak"))

d <- "~/Mouse_Cereb/006_peaks/001_peak_annotation/liftover_vertebrates/"

f <- list.files(d, pattern = ".bed")

lapply(f, function(x) {
  bed <- read_tsv(paste0(d, x), col_names = F)
  conserved <- bed$X4
  species <- gsub("mouse_", "", x)
  species <- gsub(".bed", "",species)
  o <- data.frame(peak=mouse_peaks$peak, species=mouse_peaks$peak %in% conserved, stringsAsFactors = F)
  colnames(o)[2] <- species
  mouse_peaks <<- left_join(mouse_peaks, o)
  return(NULL)
})

mouse_peaks <- select(mouse_peaks, - chr, -start, -end)

mouse_peaks <- mouse_peaks[, c("peak", "rat", "guinea_pig", "rabbit", "human", "rhesus", "marmoset", "dog", "cow", "elephant", "opossum", "platypus", "chicken", "zebra_finch", "anolis", "xenopus", "zebrafish", "medaka", "lamprey")]

plot(apply(mouse_peaks[, 2:ncol(mouse_peaks)], 2, sum), log = "y")
```

Let's also check the agreement across species with the same phylogenetic distance

```{r}
## Vertebrate conserved
sum(mouse_peaks$zebrafish)
sum(mouse_peaks$medaka)

sum(mouse_peaks$zebrafish & mouse_peaks$medaka)
sum(mouse_peaks$zebrafish & mouse_peaks$medaka)/sum(mouse_peaks$zebrafish)

## Amniote conserved
sum(mouse_peaks$zebra_finch)
sum(mouse_peaks$chicken)
sum(mouse_peaks$anolis)

sum(mouse_peaks$zebra_finch & mouse_peaks$chicken & mouse_peaks$anolis)
sum(mouse_peaks$zebra_finch & mouse_peaks$chicken)/sum(mouse_peaks$zebra_finch)

## Euarchontaglire conserved
sum(mouse_peaks$human)
sum(mouse_peaks$rhesus)
sum(mouse_peaks$marmoset)

sum(mouse_peaks$human & mouse_peaks$rhesus & mouse_peaks$marmoset)

sum(mouse_peaks$human & mouse_peaks$rhesus & mouse_peaks$marmoset)/sum(mouse_peaks$human)
```

Let's proceed with assigning ages to the different regions.
This is based on the age of the DNA sequence, not the actual regulatory element.

```{r}
mouse_peaks$min_age <- factor(ifelse(mouse_peaks$lamprey, yes = ">600", 
                              no = ifelse(mouse_peaks$zebrafish | mouse_peaks$medaka, yes = "435", 
                                          no = ifelse(mouse_peaks$xenopus, yes = "350",
                                                      no = ifelse(mouse_peaks$anolis | mouse_peaks$zebra_finch | mouse_peaks$chicken, yes = "312",
                                                                  no = ifelse(mouse_peaks$platypus, yes="177",
                                                                              no = ifelse(mouse_peaks$opossum, yes = "160",
                                                                                          no = ifelse(mouse_peaks$elephant, yes = "105",
                                                                                                      no = ifelse(mouse_peaks$dog | mouse_peaks$cow, yes = "96",
                                                                                                                  no = ifelse(mouse_peaks$human | mouse_peaks$rhesus | mouse_peaks$marmoset, yes = "90", 
                                                                                                                              no = ifelse(mouse_peaks$rabbit, yes = "82", 
                                                                                                                                          no = ifelse(mouse_peaks$guinea_pig, yes = "73", 
                                                                                                                                                      no = ifelse(mouse_peaks$rat, yes = "20", no = "<20")))))))))))), levels = c("<20", "20", "73", "82", "90", "96", "105", "160", "177", "312", "350", "435", ">600"))

table(mouse_peaks$min_age)

mouse_peaks$nSpecies <- apply(mouse_peaks[, 2:19], 1, sum)

p <- ggplot(mouse_peaks, aes(x=min_age, y=nSpecies, fill=min_age)) +
  geom_boxplot(outlier.alpha = 0) +
  scale_fill_grey(start = 0.9, end = 0.1) +
  xlab("Minimum age") +
  ylab("Number of species") +
  theme_classic() +
  theme(legend.position = "none")

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_minimumAge_and_numberOfSpecies.pdf", width = 6, height = 4, useDingbats = F); print(p); dev.off()
```

We can proceed with this although it will probably be more meaningful to group our regions in age group bins.

```{r}
ggplot(mouse_peaks, aes(min_age)) +
  geom_bar() +
  theme_classic()
```

Generating a collapsed annotation:

We will split in the following classes
1. <= 20 Mya (muridae)
2. <= 105 Mya (eutherian)
3. <= 177 Mya (mammalian)
3. <= 350 Mya (amniote)
4. >=435 105 Mya (vertebrate)

```{r}
ages.df <- data.frame(min_age=c("<20", "20", "73", "82", "90", "96", "105", "160", "177", "312", "350", "435", ">600"), min_age_collapsed=c("0-20_muridae", "0-20_muridae", "73-105_eutherian", "73-105_eutherian", "73-105_eutherian", "73-105_eutherian", "73-105_eutherian", "160-177_mammalian", "160-177_mammalian", "312-350_amniote", "312-350_amniote", "435_vertebrate", "435_vertebrate"), min_age_num=c(0, 20, 73, 82, 90, 96, 105, 160, 177, 312, 350, 435, 600), stringsAsFactors = F)

mouse_peaks <- left_join(mouse_peaks, ages.df)
```

Finally, let's put this together with the rest of the annotation:

```{r}
peaks <- left_join(peaks, select(mouse_peaks, peak, min_age, min_age_num, min_age_collapsed, nSpecies))
peaks$min_age <- factor(peaks$min_age, levels = c("<20", "20", "73", "82", "90", "96", "105", "160", "177", "312", "350", "435", ">600"))
```

```{r}
p <-ggplot(peaks, aes(genomic_class, fill=min_age)) +
  geom_bar(position = "fill") +
  theme_classic() +
  scale_fill_manual(values = colorRampPalette(brewer.pal(11, "Spectral"))(length(unique(peaks$min_age))), name="Minimum\nage") +
  theme(axis.text.x = element_text(angle=45, hjust = 0.95))

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_minimumAge_byGenomicClass.pdf", width = 8, height = 5, useDingbats = F); print(p); dev.off()
```

```{r}
p <- ggplot(peaks, aes(x=min_age, y=meanPhastCons_100bp, fill=min_age)) +
  geom_boxplot(notch = T,width=0.8, outlier.alpha = 0) +
  scale_fill_manual(values = colorRampPalette(brewer.pal(11, "Spectral"))(length(unique(peaks$min_age))), guide=F) +
  theme_classic() +
  ylab("PhastCons (vertebrates)") +
  xlab("Minimum age") 

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_phastCons_vertebrates_byMinimumAge.pdf", width = 5, height = 4, useDingbats = F); print(p); dev.off()
```

```{r}
p <- ggplot(peaks, aes(x=min_age, y=meanPhastCons_Glires_100bp, fill=min_age)) +
  geom_boxplot(notch = T,width=0.8, outlier.alpha = 0) +
  scale_fill_manual(values = colorRampPalette(brewer.pal(11, "Spectral"))(length(unique(peaks$min_age))), guide=F) +
  theme_classic() +
  ylab("PhastCons (glires)") +
  xlab("Minimum age") 

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_phastCons_glires_byMinimumAge.pdf", width = 5, height = 4, useDingbats = F); print(p); dev.off()
```


## Repeats

Now we can intersect our peaks with repeat regions (downloaded from the UCSC Table Browser [mm10/Variartion and Repeats/RepeatMasker/rmsk] on 09.09.2019)

First we need to convert the table into a bed-like file:

```{r}
repeatmasker.info <- read_tsv("006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker.txt")

head(repeatmasker.info)

## Probably the safest way to deal with this data is to generate unique identifiers
repeatmasker.info$repeatElement <- paste("Rep", repeatmasker.info$genoName, repeatmasker.info$genoStart, repeatmasker.info$genoEnd, sep = "_")

length(unique(repeatmasker.info$repeatElement))/nrow(repeatmasker.info)

## The only non-unique element is chr9_119078472_119078571 which matches two ERV elements.

## Let's make a bed file
repeatmasker.bed <- select(repeatmasker.info, genoName, genoStart, genoEnd, repeatElement)
write_tsv(repeatmasker.bed, "006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker.bed", col_names = F)

write_tsv(repeatmasker.info, "006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker_info.txt", col_names = T)
```

```{bash}
bedtools intersect -a 006_peaks/Mouse_Cerebellum_reproducible_peaks.bed -b 006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker.bed -wo > 006_peaks/001_peak_annotation/repeatmasker/Mouse_Cerebellum_peaks_overlap_repeatmasker.txt
```

Now let's read in the overlaps:

```{r}
repeat.overlaps <- read_tsv("006_peaks/001_peak_annotation/repeatmasker/Mouse_Cerebellum_peaks_overlap_repeatmasker.txt", col_names = F)

repeatmasker.info <- read_tsv("006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker_info.txt")

repeat.overlaps <- mutate(repeat.overlaps, width= (X3 - X2)) %>%
  select(width, X4, X8, X9) %>%
  rename(peak=X4, repeatElement=X8, repeatOverlap=X9) %>%
  left_join(select(repeatmasker.info, repeatElement, repName, repClass, repFamily))

repeat.overlaps$repFraction <- repeat.overlaps$repeatOverlap / repeat.overlaps$width

ggplot(repeat.overlaps, aes(log10(repeatOverlap))) +
  geom_histogram()

ggplot(repeat.overlaps, aes(repFraction)) +
  geom_histogram()

```

We see that the size of the average overlap is rather fixed (at around 100 bp).

However at the moment we are considering each overlap separately. Thus, it's probably wiser to simply focus on the ID of overlapping repeats.

```{r}
repeat.overlaps_byPeak <- group_by(repeat.overlaps, peak) %>%
  summarise(overlappingRepElements=paste(repeatElement, collapse = ","),
            overlappingRepName=paste(repName, collapse = ","),
            overlappingRepFamily=paste(repFamily, collapse = ","),
            overlappingRepClass=paste(repClass, collapse = ","))

nrow(repeat.overlaps_byPeak)
nrow(repeat.overlaps)

peaks <- left_join(peaks, repeat.overlaps_byPeak)
```

We can estimate the fraction of each peak covered by repeats by collapsing the repeat bed file and then overlapping with that.

```{bash}
bedtools sort -i 006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker.bed | bedtools merge | awk '{print $1 "\t" $2 "\t" $3 "\t" "mergedRep_" $1 "_" $2 "_" $3}' > 006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker_nonRedundant.bed

bedtools intersect -a 006_peaks/Mouse_Cerebellum_reproducible_peaks.bed -b 006_peaks/001_peak_annotation/repeatmasker/mm10_Repeatmasker_nonRedundant.bed -wo > 006_peaks/001_peak_annotation/repeatmasker/Mouse_Cerebellum_peaks_overlap_repeatmasker_nonRedundant.txt
```

```{r}
repeat.overlaps_nr <- read_tsv("006_peaks/001_peak_annotation/repeatmasker/Mouse_Cerebellum_peaks_overlap_repeatmasker_nonRedundant.txt", col_names = F)

repeat.overlaps_nr_sum <- select(repeat.overlaps_nr, X4, X9) %>%
  rename(peak=X4) %>%
  group_by(peak) %>%
  summarise(repeatOverlap=sum(X9))

peaks <- left_join(peaks, repeat.overlaps_nr_sum)
peaks$repeatOverlap[is.na(peaks$repeatOverlap)] <- 0

ggplot(peaks, aes(log10(repeatOverlap + 1))) +
  geom_histogram(bins = 50)

peaks$repeatFraction <- peaks$repeatOverlap / 500

ggplot(peaks, aes(repeatFraction)) +
  geom_histogram(bins = 50)
```

## CRE Assignment to target genes

Here we will use the results of dataset-wide correlation between peak activity and inferred gene expression (based on ArcR's gene score matrix) to assign CREs to their target genes.

```{r}
peak2gene_corrs <- readRDS("~/Mouse_Cereb/009_peak2gene/Mouse_Cerebellum_peak2gene_upTo250Kb_all.rds")

hist(peak2gene_corrs$Correlation, breaks = 100)
abline(v = 0.4, col="red", lty="dashed")

max(peak2gene_corrs$FDR[peak2gene_corrs$Correlation > 0.4])
```

Establishing empirical cutoffs based on correlation across different chromosomes:

```{r}
## 1,000 peaks from chr1
set.seed(1)
atac_knn <- readRDS(peak2gene_corrs@metadata$seATAC)
atac.chr1 <- sample(atac_knn@rowRanges[seqnames(atac_knn@rowRanges) == "chr1"], 1e3)
atac.chr1.knn <- assay(atac_knn[atac_knn@rowRanges %in% atac.chr1])
atac.chr1.knn <- atac.chr1.knn[rowSums(atac.chr1.knn) > 0,]
dim(atac.chr1.knn)

## 1,000 genes from chr2, chr3 and chr4
rna_knn <- readRDS(peak2gene_corrs@metadata$seRNA)
rna.chr234 <- sample(rna_knn@rowRanges[seqnames(rna_knn@rowRanges) %in% c("chr2", "chr3", "chr4")], 1e3)
rna.chr234.knn <- assay(rna_knn[rna_knn@rowRanges %in% rna.chr234])
rna.chr234.knn <- rna.chr234.knn[rowSums(rna.chr234.knn) > 0,]
dim(rna.chr234.knn)

null_cors <- cor(t(atac.chr1.knn), t(rna.chr234.knn))
hist(as.numeric(null_cors), breaks = 100)
quantile(as.numeric(null_cors), 0.99)
```

```{r}
dim(atac.chr1.knn)
dim(rna.chr234.knn)

row.names(atac.chr1.knn[3,])

null_cors <- cor(t(atac.chr1.knn), t(rna.chr234.knn))
hist(as.numeric(null_cors), breaks = 100)

quantile(as.numeric(null_cors), 0.99)
```

Putting them all together for a plot

```{r}
set.seed(1)
corr.df <- data.frame(corr=c(sample(peak2gene_corrs$Correlation, 1e5), sample(as.numeric(null_cors), 1e5)), type=factor(c(rep("250Kb window", 1e5), rep("Different chrom.", 1e5))))

p <- ggplot(corr.df, aes(corr, color=type, fill=type)) +
  geom_density(alpha=0.3) +
  scale_color_manual(values = c("forestgreen", "mediumorchid")) +
  scale_fill_manual(values = c("forestgreen", "mediumorchid")) +
  geom_vline(xintercept = 0.45, color="red", lty="dashed") +
  annotate(geom="text",x=0.65, y=2, label="r=0.45\nFDR<0.01", color="red") +
  xlab("Pearson's correlation") +
  theme_classic()

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_assignment_toGenes_nullDistribution.pdf", width = 6, height = 4, useDingbats = F); print(p); dev.off()
```

Now selecting significant correlations and assigning peaks to genes

```{r}
peak2gene <- data.frame(peak=peak2gene_corrs@metadata$peakSet[peak2gene_corrs$idxATAC[peak2gene_corrs$Correlation > 0.45]],
                        target_gene=peak2gene_corrs@metadata$geneSet[peak2gene_corrs$idxRNA[peak2gene_corrs$Correlation > 0.45]],
                        target_gene_corr=peak2gene_corrs$Correlation[peak2gene_corrs$Correlation > 0.45],
                        stringsAsFactors = F) %>%
  mutate(peak=paste(peak.seqnames, peak.start, peak.end, sep = "_")) %>%
  rename(target_gene=target_gene.name,
         target_gene_TSS=target_gene.start) %>%
  group_by(peak,target_gene) %>%
  mutate(target_gene_dist=min(abs(target_gene_TSS-peak.start), abs(target_gene_TSS-peak.end)),
         target_gene_TSSpos=ifelse(target_gene_TSS <= peak.end & target_gene_TSS >= peak.start, yes="overlap", 
                                   no = ifelse(target_gene_TSS < peak.start, yes = "upstream", no = "downstream"))) %>%
  ungroup() %>%
  select(peak, target_gene, target_gene_corr, target_gene_TSS, target_gene_TSSpos, target_gene_dist)
  
nrow(peak2gene)

length(unique(peak2gene$peak))

summary(as.numeric(table(peak2gene$peak)))
quantile(as.numeric(table(peak2gene$peak)), 0.95)
sum(as.numeric(table(peak2gene$peak))==1)
```

We have associated 83K peaks with a gene. 50% of them are only associated with a single gene.

For now, let's collapse to the peak level, but also export this table separately.

```{r}
peak.targets <- group_by(peak2gene, peak) %>%
  summarise(put_target_genes=paste(target_gene, collapse = ","))

head(peak.targets)

peaks <- left_join(peaks, peak.targets)
```

```{r}
p <- ggplot(peaks, aes(genomic_class, fill=is.na(put_target_genes))) +
  geom_bar(position = "fill") +
  theme_classic() +
  theme(axis.text.x = element_text(angle=45, hjust = 0.95))

p
```

## Activity in dataset and robust peak identification

We have previously seen that our sequencing depth is so high that allows us to call peaks even for regions with noisy and very sparse activity. This is especially prevelent in abundant cell types (early progenitors, mature GCs). We should limit our analyses to peaks with robust activity.

First we need to profile the activity across the dataset. For that we will use our 1st level clustering (C47):

```{r}
peak_mat <- readRDS("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peak_matrix_SE.rds")
```

A serious issue here is that the order of the peaks is mixed. In the archR project, seqnames (chromosomes) are treated as factors, thus we go with the order chr1, chr2, chr3 ... chrX

However, when creating the summarised experiment, seqnames are first converted into characters, thus ordered as chr1, chr10, chr11, chr12, ... chr19, chr2, chr3 ... chrX, and then are lost.

We can identify this issue by plotting the indexes of the peaks

Indexes in peak info:

```{r}
unique(peaks$idx)
plot(peaks$idx[seq(1, length(peaks$idx), 50)])
```

Indexes in peak matrix:

```{r}
plot(as.numeric(peak_mat@elementMetadata$idx)[seq(1, nrow(peak_mat@elementMetadata$idx), 50)])
```

We can solve this by converting chr to a character, reordering the peaks based on chr, index, then adding the peak names to the peak matrix.

```{r}
sum(peaks$idx == peak_mat@elementMetadata$idx)

peaks <- mutate(peaks, chr=as.character(chr)) %>%
  arrange(chr, idx)

sum(peaks$idx == peak_mat@elementMetadata$idx)
```

Now we can add everything stored in peakset info as rowData of peak mat

```{r}
rowData(peak_mat)

rowData(peak_mat) <- peaks

peak_mat@elementMetadata$peak[1:10]
```

Now we will estimate the frequency of cells within a cluster that show accessibility for each peak.

We have previously seen that requiring at least 5% of the cells in at least one cluster to show activity is a good way to identify high confidence peaks, but we can reevaluate this cutoff now.

```{r}
clust <- sort(unique(peak_mat@colData$Clusters_finalNucSet_res1.5))

peak_clust_f <- do.call(cbind, lapply(clust, function(id) {
  cell_i <- which(peak_mat@colData$Clusters_finalNucSet_res1.5==id)
  return(Matrix::rowSums(assay(peak_mat[,cell_i]) > 0)/length(cell_i))
}))

colnames(peak_clust_f) <- clust
row.names(peak_clust_f) <- peak_mat@elementMetadata$peak
```

```{r}
peaks$max_freq <- apply(peak_clust_f, 1, max)

sum(peaks$max_freq >= 0.01)
sum(peaks$max_freq >= 0.05)
sum(peaks$max_freq >= 0.1)

ggplot(peaks, aes(log10(max_freq))) +
  geom_histogram(bins = 30) +
  theme_classic()
```

Let's see how peaks passing a cutoff of 5% would compare to the ones we want to exclude from downstream analyses:

```{r}
p <- ggplot(peaks, aes(x=is.na(put_target_genes)==F, y=log10(max_freq), fill=as.factor(is.na(put_target_genes)==F))) +
  geom_violin() +
  geom_boxplot(notch = T, width=0.1) +
  scale_fill_manual(values = c("gray70", "forestgreen"), guide=F) +
  theme_classic() +
  xlab("Assigned to target gene") +
  geom_hline(yintercept = log10(0.05), color="red", lty="dashed")

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_maxFreq_byTargetGene.pdf", width = 2.5, height = 3, useDingbats = F); print(p); dev.off()
```

```{r}
p <- ggplot(peaks, aes(x=log10(max_freq), y=log10(score))) +
  geom_hex() +
  theme_classic()

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_maxFreq_byPeakScore.pdf", width = 5, height = 4, useDingbats = F); print(p); dev.off()
```

```{r}
p <- ggplot(peaks, aes(x=as.factor(Reproducibility), y=log10(max_freq), fill=as.factor(Reproducibility))) +
  geom_violin() +
  geom_boxplot(notch = T, width=0.1) +
  scale_fill_brewer(palette = "Greens", name="") +
  geom_hline(yintercept = log10(0.05), color="red", lty="dashed") +
  xlab("Number of samples supporting peak") +
  theme_classic()

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_maxFreq_byReproducibility.pdf", width = 5, height = 3, useDingbats = F); print(p); dev.off()
```

The cutoff of 0.05 seems quite reasonable. Let's proceed with this:

```{r}
cutoff <- 0.05

peaks$robust <- peaks$max_freq >= cutoff 

p <- ggplot(data = NULL, aes(x=seq(1, length(peaks$max_freq), 20),y=sort(peaks$max_freq[seq(1, length(peaks$max_freq), 20)], decreasing = T), color=sort(peaks$max_freq[seq(1, length(peaks$max_freq), 20)], decreasing = T) >= cutoff )) +
  geom_point() +
  geom_hline(yintercept = cutoff , color="red", lty="dashed") +
  geom_vline(xintercept = which.min(sort(peaks$max_freq, decreasing = T) >= cutoff ), color="red", lty="dashed") +
  scale_color_manual(values = c("gray70", "deepskyblue3"), guide=F) +
  annotate(geom="text",x=which.min(sort(peaks$max_freq, decreasing = T) >= cutoff ) - 5e4, y=0.2, label=paste0("Robust peaks\nFreq >= ",cutoff ,"\nN=", which.min(sort(peaks$max_freq, decreasing = T) >= cutoff )), color="deepskyblue3") +
  ylab("Maximum fraction of active cells in a cluster") +
  xlab("Peaks (decreasing maximum activity)") +
  theme_classic()

p

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_robustPeak_identification.pdf", width = 6, height = 4, useDingbats = F); print(p); dev.off()
```

## Benchmarking with external datasets

## Vista Enhancers

Downloading positive VISTA enhancers (31.07.19), lifting over to mm10 and summarising (see details in the associated script).

Now we just want to intersect with our peaks:

```{bash}
bedtools intersect -a 006_peaks/Mouse_Cerebellum_reproducible_peaks.bed -b 006_peaks/001_peak_annotation/vista/190731_VISTA_mouse_positive_mm10.bed -wa -wb > 006_peaks/001_peak_annotation/vista/Mouse_Cerebellum_peaks_overlap_vista.txt
```

```{r}
vista.overlaps <- read_tsv("006_peaks/001_peak_annotation/vista/Mouse_Cerebellum_peaks_overlap_vista.txt", col_names = F)

## VISTA elements overlapping a robust peak in our dataset
vista.info <- read_tsv("006_peaks/001_peak_annotation/vista/190731_VISTA_mouse_positive_summary.txt")
head(vista.info)

vista.info.gather <- separate_rows(vista.info, activity, sep = ",") %>%
  mutate(in_scATAC=mm9_region %in% vista.overlaps$X8[vista.overlaps$X4 %in% peaks$peak[peaks$robust]])

table(vista.info.gather$in_scATAC)

vista_byTissue <- group_by(vista.info.gather, activity) %>%
  summarise(total_found = sum(in_scATAC), fraction_found = sum(in_scATAC) / length(in_scATAC)) %>%
  filter(total_found >=20) %>%
  arrange(desc(fraction_found)) %>%
  mutate(activity=factor(activity, levels = unique(activity)))


p <- ggplot(vista_byTissue, aes(x=activity, y=fraction_found)) +
  geom_col() +
  theme_classic() +
  xlab("Vista activity") +
  ylab("Fraction of vista elements in robust peaks") +
  theme(axis.text.x = element_text(angle=45, hjust=0.99))

p 

pdf("~/Mouse_Cereb/Figures/003_peakAnnotation/Mouse_peaks_vistaEnrichmentByTissue_inRobustPeaks.pdf", width = 6, height = 4, useDingbats = F); print(p); dev.off()

## By peak
vista.overlaps_byPeak <- select(vista.overlaps, X4, X8) %>%
  dplyr::rename(peak=X4, mm9_region=X8) %>%
  left_join(vista.info) %>%
  group_by(peak) %>%
  summarise(vista_element=paste(element, collapse = "|"),
            vista_activity=paste(activity, collapse = "|"))

nrow(vista.overlaps_byPeak)

peaks <- left_join(peaks, vista.overlaps_byPeak)

ggplot(peaks, aes(as.factor(is.na(vista_activity)==F), fill=robust)) +
  geom_bar(position = "fill") +
  xlab("Vista activity")
```

## Export peak stats

```{r}
write_tsv(peaks, "~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peaks_summary.txt")
write_tsv(peak2gene, "~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peaks_assignment_to_Genes.txt")
## Updating metadata on matrix and exporting
rowData(peak_mat) <- peaks
saveRDS(peak_mat, "~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peak_matrix_SE_withInfo.rds")
saveRDS(peak_mat[peak_mat@elementMetadata$robust==T], "~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peak_matrix_SE_withInfo_robustPeaks.rds")
```


```{r}
sessionInfo()
```

