---
title: "Mouse Cerebellum - 004 TF activity across pseudotime"
author: "Ioannis Sarropoulos"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(Mfuzz)
  library(RColorBrewer)
  library(viridis)
  library(ComplexHeatmap)
})
```

A general problem with such analyses is that a lot of tfs with similar motifs will show similar activity.
This is why we will use our inferred gene activity to refine our analysis to TFs with:

1. Dynamic motif
2. Dynamic RNA (imputed through gene activity)
3. (Absolute) correlation between motif activity and RNA levels
4. Reaching a minimum gene activity score in at least one bin.

```{r}
#1. Which lineage is to be processed?
lineage <- "PC"
#2. Correlation cutoff (absolute)
corr_cutoff <- 0.6
#3. Gene activity cutoff (minimum gene expression activity to consider TF active)
rna_cutoff <- 1
#4. Optimal number of Mfuzz clusters for TFs: Determined by calculating distance between cluster centroids, visual observation of trajectories, and GO enrichment of each cluster.
N_tf_clusters <- 3
```

Load data

```{r}
pseudotime_bins <- read_tsv(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_pseudotime_stats.txt"))

## Motif activity matrix
motifs <- read.table(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_lineage_pseudotime_bins_chromVar_deviations_pseudobulk.txt"), header = T, sep = "\t", stringsAsFactors = F)

## Gene score activity matrix
genes <- read.table(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_lineage_pseudotime_bins_geneScoresScaled_pseudobulk.txt"), header = T, sep = "\t", stringsAsFactors = F)

## Motif MMI
motifs_mmi <- read_tsv(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_lineage_motifs_pseudotime_MMI.txt"))

## Genes MMI
genes_mmi <- read_tsv(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_lineage_genes_pseudotime_MMI.txt"))

## Significant genes
sig.genes <- readLines(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_lineage_genes_pseudotime_MMI_significant.txt"))

## From which we infer the MMI cutoff
mmi_cutoff <- round(min(genes_mmi$mmi[genes_mmi$gene %in% sig.genes]), digits = 3)
mmi_cutoff

sum(genes_mmi$mmi > mmi_cutoff)
sum(motifs_mmi$mmi > mmi_cutoff)
```

Collecting statistics for the motif and imputed RNA, as well as the correlation between motif activity and gene score.

```{r}
tf_activity <- separate(motifs_mmi, motif, into = c("gene", "n"), sep = "-",remove = F) %>%
  dplyr::rename(motif_mmi=mmi, motif_pval=pval, motif_padj=padj) %>%
  inner_join(genes_mmi) %>%
  dplyr::rename(rna_mmi=mmi, rna_pval=pval, rna_padj=padj)

tf_activity$corr_motif_rna <- sapply(1:nrow(tf_activity), function(i) {
    cor(as.numeric(motifs[tf_activity$motif[i], ]), as.numeric(genes[tf_activity$gene[i], ]))
})

hist(tf_activity$corr_motif_rna, breaks = 20)
```

```{r}
tf_activity$rna_max <- sapply(1:nrow(tf_activity), function(i) {
    max(as.numeric(genes[tf_activity$gene[i], ]))
})

hist(log10(tf_activity$rna_max), breaks = 30)
plot(sort(tf_activity$rna_max))
```

```{r}
plot(tf_activity$rna_max, tf_activity$rna_mmi)
plot(tf_activity$rna_max, tf_activity$corr_motif_rna)
```

```{r}
ggplot(tf_activity, aes(x=rna_mmi, y=motif_mmi, color=corr_motif_rna)) +
  geom_point() +
  scale_color_gradient2(low = "red", mid = "white", high = "blue", midpoint = 0) +
  geom_vline(xintercept = mmi_cutoff, color="black", lty="dashed") +
  geom_hline(yintercept = mmi_cutoff, color="black", lty="dashed") +
  theme_bw()

ggplot(tf_activity, aes(x=rna_mmi, y=motif_mmi, color=corr_motif_rna)) +
  geom_point() +
  scale_color_gradient2(low = "red", mid = "white", high = "blue", midpoint = 0) +
  geom_text(aes(label=gene), check_overlap = T,  size=4, color="black")

ggplot(tf_activity, aes(x=corr_motif_rna, y=rna_mmi, color=corr_motif_rna, size=rna_max)) +
  geom_point() +
  scale_color_gradient2(low = "red", mid = "gray90", high = "blue", midpoint = 0)+
  theme_bw()
```

```{r}
sig.tfs <- filter(tf_activity,
                  rna_mmi > mmi_cutoff,
                  motif_mmi > mmi_cutoff, 
                  abs(corr_motif_rna) > corr_cutoff,
                  rna_max >= rna_cutoff)%>%
  arrange(motif)

sig.tfs$class <- ifelse(sig.tfs$corr_motif_rna > 0, yes = "activator", no = "repressor")

table(sig.tfs$class)
```

Now we can take all these TFs and cluster them based on their activity across pseudotime

```{r}
motifs_tfs <- as.matrix(motifs[sig.tfs$motif,])

motifs_tfs.z <- motifs_tfs - apply(motifs_tfs, 1, mean)
motifs_tfs.z <- motifs_tfs.z / apply(motifs_tfs, 1, sd)
```

Mfuzz clustering

```{r}
motifs_tfs.z.eset <- ExpressionSet(motifs_tfs.z)

m1 <- mestimate(motifs_tfs.z.eset)
m1

## So far, maximising the minimum distance between cluster centroids seems to be the best way to optimise the number of clusters.
set.seed(1)
c_dmin_tfs <- Dmin(motifs_tfs.z.eset,m1,crange=seq(3,10,1),repeats = 5)

set.seed(1)
tfs_clust <- mfuzz(motifs_tfs.z.eset,c=N_tf_clusters,m=m1)

mfuzz.plot2(motifs_tfs.z.eset, tfs_clust, colo = "fancy", x11 = F, centre = T, mfrow = c(2,3))
table(tfs_clust$cluster)
```

```{r, fig.height=6, fig.width=8}
## Counting timepoints abundance per pseudotime bin
bin_counts <- group_by(pseudotime_bins, pseudotime_bin, timepoint) %>%
  dplyr::count() %>%
  spread(key=timepoint, value=n, fill = 0) %>%
  as.data.frame()
row.names(bin_counts) <- bin_counts$pseudotime_bin
bin_counts <- as.matrix(bin_counts[,2:ncol(bin_counts)])

timepoint_cols <- RColorBrewer::brewer.pal(11, "Spectral")
names(timepoint_cols) <- c("e10", "e11", "e12", "e13", "e15", "e17", "P0", "P04", "P07", "P14", "P63")
## Mean pseudotime per bin
pseudotime_mean <- group_by(pseudotime_bins, pseudotime_bin) %>%
  summarise(pseudotime=mean(pseudotime))

col_fun = circlize::colorRamp2(c(0, 1), c("white", "darkblue"))

ha = HeatmapAnnotation(stage = anno_barplot(bin_counts, gp = gpar(fill = timepoint_cols[colnames(bin_counts)], lwd=0),bar_width = 1, height = unit(1, "cm")),
                       pseudotime=pseudotime_mean$pseudotime, border = TRUE, col=list(pseudotime=col_fun), show_legend = F)



## Adding info on tf family and correlation betwenen motif and (imputed rna)
tf_family <- read_tsv("~/Resources/cisBP/TF_Information.txt") %>%
  dplyr::select(TF_Name, Family_Name) %>%
  unique() %>%
  dplyr::rename(gene=TF_Name, TF_family=Family_Name) %>%
  right_join(sig.tfs)  %>%
  arrange(motif)

## Keeping only prevalent families and setting the rest to other
# top_families <- filter(tf_family, is.na(TF_family)==F) %>%
#   group_by(TF_family) %>%
#   dplyr::count() %>%
#   arrange(desc(n))

top_families <- c("AT hook", "bHLH", "bZIP", "C2H2 ZF", "Ets", "Forkhead", "Homeodomain", "Homeodomain,POU", "Nuclear receptor", "Sox", "Other")

# top_families <- top_families$TF_family[1:10]
# 
# top_families

tf_family$TF_family[!(tf_family$TF_family %in% top_families)] <- "Other"

tf_family$Mfuzz_clust <- tfs_clust$cluster

n <- brewer.pal(11, "Set3")
names(n) <- top_families

## Adding the correlation between motif and RNA
cor_col_fun = circlize::colorRamp2(c(-1, 0, 1), c("firebrick3","white", "deepskyblue3"))

left_anno <- rowAnnotation(tf_family=tf_family$TF_family, corr= tf_family$corr_motif_rna, col=list(tf_family=n, corr=cor_col_fun), border=T, annotation_legend_param = list(direction = "vertical"))

## Specifying interesting TFs to display
interesting_motifs_i <- grep("Lhx5|Neurog2|Esrrb|Olig2|Zic4|Zic5|Zeb1|Rora|Mef2c|Sox2-|Nhlh1|Foxp2|Id3|Ascl1",row.names(motifs_tfs.z))

interesting_motifs_names <- sapply(row.names(motifs_tfs.z)[interesting_motifs_i], function(x) str_split(x, "-", simplify = T)[1])

tf_names = rowAnnotation(foo = anno_mark(at = interesting_motifs_i, labels = interesting_motifs_names), annotation_legend_param = list(direction = "vertical",nrow = 1, labels_gp = gpar(fontsize = 0.1), padding = unit(10, "mm")))

draw(Heatmap(matrix = motifs_tfs.z,
        row_split =tf_family$Mfuzz_clust,
        cluster_row_slices=T,
        top_annotation = ha,
        cluster_rows = T,
        cluster_columns = F,
        clustering_distance_rows = "pearson",
        show_row_names = F,
        show_column_names = F,
        col=viridis(100,option = "A"),
        right_annotation = tf_names,
        left_annotation = left_anno,
        border = TRUE,
        heatmap_legend_param = list(direction = "vertical", title = "Z-score motif")), heatmap_legend_side = "right")

pdf(paste0("~/Mouse_Cereb/Figures/007_withinLineages/", lineage, "/Mouse_within", lineage, "_pseudotime_TFs_Mfuzz_heatmap.pdf"), width = 8, height = 8, useDingbats = F);draw(Heatmap(matrix = motifs_tfs.z,
        row_split =tf_family$Mfuzz_clust,
        cluster_row_slices=T,
        top_annotation = ha,
        cluster_rows = T,
        cluster_columns = F,
        clustering_distance_rows = "pearson",
        show_row_names = F,
        show_column_names = F,
        col=viridis(100,option = "A"),
        right_annotation = tf_names,
        left_annotation = left_anno,
        border = TRUE,
        heatmap_legend_param = list(direction = "vertical", title = "Z-score motif")), heatmap_legend_side = "right");dev.off()
```

Export

```{r}
write_tsv(tf_activity, paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_TF_motif_and_RNA_inPseudotime.txt"))

write_tsv(tf_family, paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_TF_motif_and_RNA_inPseudotime_significantTFs.txt"))
```


```{r}
sessionInfo()
```

