---
title: "Developmental peaks by cell type"
author: "Ioannis Sarropoulos"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(ComplexHeatmap)
  library(UpSetR)
  library("rGREAT")
  library(RColorBrewer)
  library(viridis)
  library(DESeq2)
})
```

### DA CREs per cell type

First, let's plot the number of DA peaks per cell type and timepoint.

We have so many DA peaks in late GCs (also due to the very high cell number and sequencing depth) that we will need to introduce breaks to the plot

```{r, fig.width=10, fig.height=12}
ct <- c("Astroglia", "GABA_DN", "Glut_DN","PC", "Interneuron", "GC")

all.gather <- Reduce(bind_rows, lapply(ct, function(x) {
  f <-readRDS(paste0("~/Mouse_Cereb/012_devPeaks/byCellType/", x, "/", x, "_pseudobulk_byTimepoint_deseq_results.rds"))
  f.gather <- gather(f, -peak, key = "timepoint", value = "sign", na.rm = F) %>%
  mutate(timepoint=factor(timepoint), sign=factor(sign)) %>%
  filter(sign %in% c("up", "down")) %>%
  group_by(timepoint, sign, .drop = FALSE) %>%
    dplyr::count() %>%
  ungroup() %>%
  filter(sign %in% c("up", "down")) %>%
    droplevels()
  f.gather$Cell_type <- x
  return(f.gather)
}))

all.gather$n[all.gather$sign=="down"] <- (-1) * all.gather$n[all.gather$sign=="down"]

all.gather$timepoint <- gsub("_", "-", all.gather$timepoint)

cuts <- c(-41000, -39000 ,-26000, -24000 ,-4000,-2000,  0, 2000, 4000,  11000,  13000,  23000,  25000)

p <- ggplot(all.gather, aes(x=Cell_type, y=n, fill=Cell_type, color=sign)) +
  geom_col(lwd=0.8) +
  facet_wrap(~timepoint, scales = "free_x", nrow = 1) +
  scale_color_manual(values=c("#ad2a31", "deepskyblue4")) +
  scale_y_continuous(breaks = cuts, labels = abs(cuts)) +
  scale_fill_manual(values = c("#a960a1", "#43bbc8", "#62943c", "#de761d","#d3405a", "#527cde")) +
  ylab("Number of DA regions") +
  xlab("Cerebellum development") +
  geom_text(aes(label=abs(n)), position = position_nudge(y = 100)) +
  theme_classic() +
  theme(axis.text.x = element_blank(), axis.ticks.x=element_blank())

p

pdf(paste0("~/Mouse_Cereb/Figures/004_devPeaks/Mouse_peaks_deseq_byStage_byCellType.pdf"), width = 10, height = 12, useDingbats = F); print(p); dev.off()
```

Now let's look if the peak sets overlap between cell types and stages:

```{r, fig.width=9, fig.height=4}
all.da <- Reduce(bind_rows, lapply(ct, function(x) {
  f <-readRDS(paste0("~/Mouse_Cereb/012_devPeaks/byCellType/", x, "/", x, "_pseudobulk_byTimepoint_deseq_results.rds"))
  f.gather <- gather(f, -peak, key = "timepoint", value = "sign", na.rm = F) %>%
  filter(sign %in% c("up", "down"))
  f.gather$Cell_type <- x
  return(f.gather)
}))

all.da$comp <- paste(all.da$Cell_type, all.da$timepoint, all.da$sign, sep = "-")

deseq_peaks <- lapply(unique(all.da$comp), function(c) {
  return(all.da$peak[all.da$comp==c])
})

names(deseq_peaks) <- unique(all.da$comp)

deseq_peaks_metadata <- data.frame(sets=names(deseq_peaks), stringsAsFactors = F) %>%
  separate(sets, into = c("Cell_type","timepoint", "sign"), sep = "-", remove = F)

p <- upset(fromList(deseq_peaks),
      order.by = "freq",
      set.metadata = list(data = deseq_peaks_metadata, plots = list(list(type = "heat", column = "sign", assign = 4, colors = c(up="deepskyblue4", down="#ad2a31")),
                                                                    list(type = "heat", column = "Cell_type", assign = 4, colors = c("#a960a1", "#43bbc8", "#62943c", "#d3405a", "#527cde")),
                                                                    list(type = "heat", column = "timepoint", assign = 4, colors = brewer.pal(10, "Spectral")[c(2,3,7, 9, 10)])
                                                                    )),
      nsets = 10,
      nintersects = 25,
      mb.ratio = c(0.5, 0.5),
      keep.order = T)

p

pdf(paste0("~/Mouse_Cereb/Figures/004_devPeaks/Mouse_peaks_deseq_byStage_byCellType_upset.pdf"), width = 9, height = 4, useDingbats = F); print(p); dev.off()

```

Characterizing these peaks using GREAT and Homer

```{r, eval=FALSE}
makeBed <- function(peaks) {
  df <- as.data.frame(do.call(rbind, lapply(peaks, function(p) {
    return(c(str_split(p, pattern = "_", simplify = T), p))
  })), stringsAsFactors=F)
  df[,2] <- as.numeric(df[,2])
  df[,3] <- as.numeric(df[,3])
  colnames(df) <- c("chr", "start", "end", "peak")
  return(df)
}

if(dir.exists("~/Mouse_Cereb/012_devPeaks/byCellType/peaks_GREAT_enrichment")==F) {
  dir.create("~/Mouse_Cereb/012_devPeaks/byCellType/peaks_GREAT_enrichment", recursive = T)
}

atac.deseq.res <- Reduce(bind_rows, lapply(ct, function(x) {
  f <-readRDS(paste0("~/Mouse_Cereb/012_devPeaks/byCellType/", x, "/", x, "_pseudobulk_byTimepoint_deseq_results.rds"))
  f.gather <- gather(f, -peak, key = "timepoint", value = "sign", na.rm = F) %>%
    filter(sign!="not-tested")
  f.gather$Cell_type <- x
  return(f.gather)
})) 

atac.deseq.res$comp <- paste(atac.deseq.res$Cell_type, atac.deseq.res$timepoint, atac.deseq.res$sign, sep = "-")

lapply(major.comp$comp, function(c) {
  x <- as.character(str_split(c, "-", simplify = T))
  cell_type <- x[1]
  timepoint <- x[2]
  sign <- x[3]
  bg <- makeBed(atac.deseq.res$peak[atac.deseq.res$Cell_type==cell_type & atac.deseq.res$timepoint==timepoint & atac.deseq.res$sign!="not-tested"])
  interest <- makeBed(atac.deseq.res$peak[atac.deseq.res$Cell_type==cell_type & atac.deseq.res$timepoint==timepoint & atac.deseq.res$sign==sign])
  print(paste0("Submitting job for ",c))
  job <- submitGreatJob(interest, bg = bg, species = "mm10", includeCuratedRegDoms = F, request_interval = 10)
  tb <- getEnrichmentTables(job, availableOntologies(job))
  print(paste0("Getting enrichment for ", c))
  saveRDS(tb, paste0("~/Mouse_Cereb/012_devPeaks/byCellType/peaks_GREAT_enrichment/Mouse_deseq_peaks_comp_", c, "_GREAT_all.rds"))
  print(paste0("Done with comp ", c))
  return(tb)
})
```

Now reading the GREAT results

```{r}
files <- list.files("~/Mouse_Cereb/012_devPeaks/byCellType/peaks_GREAT_enrichment", pattern = "Mouse_deseq_peaks_comp_")

files <- files[grepl("e13_e15", files)==F]

great_bp <- Reduce(bind_rows, lapply(files, function(f) {
  tb <- readRDS(paste0("~/Mouse_Cereb/012_devPeaks/byCellType/peaks_GREAT_enrichment/", f))
  bp <- tb[["GO Biological Process"]]
  s <- gsub("Mouse_deseq_peaks_comp_", "", f)
  s <- gsub("_GREAT_all.rds", "",s)
  bp$cell_type <- str_split(s, "-", simplify = T)[1]
  bp$t <- str_split(s, "-", simplify = T)[2]
  bp$sign <- str_split(s, "-", simplify = T)[3]
  bp$comp <- paste(bp$cell_type, bp$t, bp$sign, sep = "-")
  return(bp)
}))

#We want to keep the top results from each comparison, then collapse them to reduced categories
great_bp_top <- group_by(great_bp, comp) %>%
  top_n(10, -Hyper_Adjp_BH)

length(unique(great_bp_top$ID))

#write(unique(great_bp_top$ID), "~/Mouse_Cereb/012_devPeaks/byCellType/peaks_GREAT_enrichment/Mouse_GREAT_BP_top10_redundant.txt")

## Summarising in Revigo (http://revigo.irb.hr/) with the following parameters:
## Medium list (0.7)
## Database: Mus musculus
## similarity: SimRel

## Citation
#Supek F, Bošnjak M, Škunca N, Šmuc T.
#"REVIGO summarizes and visualizes long lists of Gene Ontology terms"
#PLoS ONE 2011. doi:10.1371/journal.pone.0021800

great_bp_revigo <- read_csv("~/Mouse_Cereb/012_devPeaks/byCellType/peaks_GREAT_enrichment/Mouse_GREAT_BP_top10_revigo.csv")

## Filtering BP enrichment for terms in REVIGO
great_bp_plot <- filter(great_bp, ID %in% great_bp_revigo$term_ID[great_bp_revigo$eliminated==0])

## Making a spread dataframe
great_bp_plot_spread <- dplyr::select(great_bp_plot, ID, name, Hyper_Adjp_BH, comp) %>%
  spread(key = comp, value = Hyper_Adjp_BH) %>%
  as.data.frame()

great_bp_plot_spread[,3:ncol(great_bp_plot_spread)] <- -log10(great_bp_plot_spread[,3:ncol(great_bp_plot_spread)])

## Capping at 10
great_bp_plot_spread[,3:ncol(great_bp_plot_spread)][great_bp_plot_spread[,3:ncol(great_bp_plot_spread)] > 10] <- 10

row.names(great_bp_plot_spread) <- str_trunc(great_bp_plot_spread$name, 70, "right")
```

```{r}
bp_cell_type <- sapply(colnames(great_bp_plot_spread[,3:ncol(great_bp_plot_spread)]), function(x) {
  cell <- str_split(x, "-", simplify = T)[1]
})

bp_cell_type

bp_cell_type_cols <- c("#a960a1", "#43bbc8", "#62943c", "#de761d","#d3405a", "#527cde")
names(bp_cell_type_cols) <- sort(unique(bp_cell_type))

bp_timepoint <- sapply(colnames(great_bp_plot_spread[,3:ncol(great_bp_plot_spread)]), function(x) {
  t <- str_split(x, "-", simplify = T)[2]
})

bp_timepoint_cols <- RColorBrewer::brewer.pal(length(unique(bp_timepoint)), "Spectral")
names(bp_timepoint_cols) <- unique(bp_timepoint)

bp_sign <- sapply(colnames(great_bp_plot_spread[,3:ncol(great_bp_plot_spread)]), function(x) {
  s <- str_split(x, "-", simplify = T)[3]
})

bp_sign_cols <- c("#ad2a31", "deepskyblue4")
names(bp_sign_cols) <- c("down", "up")

bp_ha = HeatmapAnnotation(cell_type = bp_cell_type, stage=bp_timepoint, sign=bp_sign,col=list(cell_type=bp_cell_type_cols, stage=bp_timepoint_cols, sign=bp_sign_cols),annotation_name_side = "left", annotation_legend_param = list(direction = "horizontal",nrow = 2))

interesting_terms <- c("cell differentiation", "neuron differentiation", "cell fate commitment", "axon midline choice point recognition", "embryonic organ development", "cell adhesion", "neuron recognition", "regulation of neuron migration", "ion transport", "membrane hyperpolarization", "steroid metabolic process", "ammonium transmembrane transport", "neurotransmitter receptor metabolic process", "ERBB3 signaling pathway", "RNA splicing", "intrinsic apoptotic signaling pathway", "regulation of activation of Janus kinase activity", "miRNA catabolic process", "chemical synaptic transmission", "response to growth factor", "response to glucose")

great_ha = rowAnnotation(foo = anno_mark(at = which(great_bp_plot_spread$name %in% interesting_terms), labels = str_trunc(great_bp_plot_spread$name[which(great_bp_plot_spread$name %in% interesting_terms)], 48, "right")), annotation_legend_param = list(direction = "vertical",nrow = 1, labels_gp = gpar(fontsize = 0.1), padding = unit(10, "mm")))
```

```{r, fig.width=10, fig.height=10}
print(draw(Heatmap(matrix = as.matrix(great_bp_plot_spread[,3:ncol(great_bp_plot_spread)]),
        cluster_rows = T,
        cluster_columns = T,
        show_column_names = T,
        show_row_names = F,
        clustering_distance_rows = "pearson",
        clustering_distance_columns = "pearson",
        clustering_method_columns = "ward.D2", 
        clustering_method_rows = "ward.D2",
        col=viridis(100,option = "D"),
        border = TRUE,
        top_annotation = bp_ha,
        right_annotation = great_ha,
        heatmap_legend_param = list( title = "-log10 P-val", direction="vertical")), heatmap_legend_side = "right", annotation_legend_side = "top"))
```

Let's also have a look into motif enrichment based on HOMER:

```{r, eval=FALSE}
if(dir.exists("~/Mouse_Cereb/012_devPeaks/byCellType/homer")==F) {
  dir.create("~/Mouse_Cereb/012_devPeaks/byCellType/homer", recursive = T)
}

lapply(major.comp$comp, function(c) {
  x <- as.character(str_split(c, "-", simplify = T))
  cell_type <- x[1]
  timepoint <- x[2]
  sign <- x[3]
  bg <- makeBed(atac.deseq.res$peak[atac.deseq.res$Cell_type==cell_type & atac.deseq.res$timepoint==timepoint & atac.deseq.res$sign!="not-tested"])
  interest <- makeBed(atac.deseq.res$peak[atac.deseq.res$Cell_type==cell_type & atac.deseq.res$timepoint==timepoint & atac.deseq.res$sign==sign])
  if(dir.exists(paste0("~/Mouse_Cereb/012_devPeaks/byCellType/homer/", c))==F) {
  dir.create(paste0("~/Mouse_Cereb/012_devPeaks/byCellType/homer/", c), recursive = T)
  }
  print(paste0("Saving bed files for ",c))
  write.table(interest, paste0("~/Mouse_Cereb/012_devPeaks/byCellType/homer/",c, "/interest.bed"), col.names = F, row.names = F, sep = "\t", quote = F)
  write.table(bg, paste0("~/Mouse_Cereb/012_devPeaks/byCellType/homer/",c, "/background.bed"), col.names = F, row.names = F, sep = "\t", quote = F)
})
```

Running Homer

```{bash, eval=FALSE}
sh ~/Mouse_Cereb/012_devPeaks/byCellType/002b_send_homer_byCellType.sh
```

Reading in the results

```{r}
motif_path <- list.dirs("~/Mouse_Cereb/012_devPeaks/byCellType/homer", recursive = F, full.names = F)

motif_path <- motif_path[grepl("e13_e15", motif_path)==F]

print(motif_path)

motifs <- Reduce(bind_rows,lapply(motif_path, function(p) {
  motifs <- read_tsv(paste0("~/Mouse_Cereb/012_devPeaks/byCellType/homer/", p, "/_Homer/knownResults.txt"))
  motifs$comp <- p
  return(motifs)
}))

motifs <-   mutate(motifs, `Motif Name`= gsub("\\/.*", "", `Motif Name`)) %>%
  group_by(`Motif Name`, comp) %>%
  summarise(`Log P-value`=min(`Log P-value`)) %>%
  ungroup()

## All motifs included in the file are significant!

motifs_spread <- dplyr::select(motifs, `Motif Name`, `Log P-value`, comp) %>%
  spread(key = comp, value=`Log P-value`)

## Now also identifying the top motif per cluster. We want to label them on the heatmap
top_motif_by_clust <- unique(apply(motifs_spread[,2:ncol(motifs_spread)], 2, function(x) motifs_spread$`Motif Name`[which.min(x)]))

top_motif_by_clust

## These aren't that informative
## Let's make our custom list

interesting_motifs <- c("Atoh1(bHLH)", "NeuroD1(bHLH)","NF1(CTF)", "CTCF(Zf)", "Lhx1(Homeobox)", "Sox2(HMG)", "ZEB1(Zf)",  "Mef2d(MADS)", "Jun-AP1(bZIP)", "RFX(HTH)", "Tgif2(Homeobox)", "Tcf12(bHLH)", "EBF2(EBF)", "Gli2(Zf)", "ZBTB18(Zf)", "Phox2a(Homeobox)", "HIC(Zf)", "COUP-TFII(NR)", "TR4(NR),DR1", "Oct4(POU,Homeobox)", "AP-2gamma(AP2)", "RORa(NR)")

## Capping to q99
motifs_spread[,2:ncol(motifs_spread)] <- motifs_spread[,2:ncol(motifs_spread)] * (-1)
motifs_spread[,2:ncol(motifs_spread)][motifs_spread[,2:ncol(motifs_spread)] > quantile(unlist(motifs_spread[,2:ncol(motifs_spread)]), 0.99)] <- quantile(unlist(motifs_spread[,2:ncol(motifs_spread)]), 0.99)

sum(apply(motifs_spread[,2:ncol(motifs_spread)], 1, max) > 10)

## Also limiting the analysis in highly significant motifs
motifs_spread.sig <- motifs_spread[apply(motifs_spread[,2:ncol(motifs_spread)], 1, max) > 10,]

## Annotating with the TF family
motif_anno <- data.frame(motif=motifs_spread.sig$`Motif Name`, stringsAsFactors = F) %>%
  mutate(tf_family=gsub("(", "", str_extract(motif, "\\(.*\\)"), fixed = T)) %>%
   mutate(tf_family=gsub(")", "", tf_family, fixed = T))

row.names(motif_anno) <- motif_anno$motif

## Keeping only prevalent families and setting the rest to other
top_families <- group_by(motif_anno, tf_family) %>%
  dplyr::count() %>%
  arrange(desc(n))

top_families <- top_families$tf_family[1:10]

motif_anno$tf_family[!(motif_anno$tf_family %in% top_families)] <- "Other"

n <- brewer.pal(11, "Set3")
names(n) <- unique(motif_anno$tf_family)
```

```{r, fig.width=8, fig.height=10}
tf_ha = rowAnnotation(foo = anno_mark(at = which(motifs_spread.sig$`Motif Name` %in% interesting_motifs), labels = motifs_spread.sig$`Motif Name`[which(motifs_spread.sig$`Motif Name` %in% interesting_motifs)]), annotation_legend_param = list(direction = "vertical",nrow = 1, labels_gp = gpar(fontsize = 0.1), padding = unit(10, "mm")))

print(draw(Heatmap(matrix = as.matrix(motifs_spread.sig[,2:ncol(motifs_spread)]),
        cluster_rows = T,
        cluster_columns = T,
        show_column_names = T,
        show_row_names = F,
        clustering_distance_rows = "pearson",
        clustering_distance_columns = "pearson",
        clustering_method_columns = "ward.D2", 
        clustering_method_rows = "ward.D2",
        col=viridis(100,option = "E"),
        border = TRUE,
        left_annotation = rowAnnotation(tf_family=motif_anno$tf_family, col=list(tf_family=n), border=T, annotation_legend_param = list(direction = "vertical")),
        right_annotation = tf_ha,
        top_annotation = bp_ha,
        heatmap_legend_param = list( title = "-log10 P-val", direction="vertical")), heatmap_legend_side = "right", annotation_legend_side = "top"))
```

### Developmental peaks at the organ level

Loading peak matrix

```{r}
peak_mat <- readRDS("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peak_matrix_SE_withInfo_robustPeaks.rds")
mouse_peaks <- read.table("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peaks_summary.txt", header = T, sep = "\t", stringsAsFactors = F)
```

Creating a pseudobulk per library

```{r}
samples <- sort(unique(peak_mat$Sample_short))

print(samples)

atac.pseudobulk <- do.call(cbind, parallel::mclapply(samples, function(s) {
  cell_i <- which(peak_mat@colData$Sample_short==s)
  atac.data <- assay(peak_mat[,cell_i])
  return(Matrix::rowSums(atac.data))
}, mc.cores = 15))

colnames(atac.pseudobulk) <- samples
rownames(atac.pseudobulk) <- peak_mat@elementMetadata$peak

saveRDS(atac.pseudobulk, "~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peak_matrix_pseudobulk_byLib.rds")
```

```{r}
dim(atac.pseudobulk)
```

```{r}
atac.info <- data.frame(sample=colnames(atac.pseudobulk), stringsAsFactors = F) %>%
  separate(sample, into = c("timepoint", "lib", "sex"), remove = F) %>%
  group_by(timepoint) %>%
  mutate(rep=1:n()) %>%
  ungroup()
```

```{r, eval=FALSE}
atac.deseq <- DESeqDataSetFromMatrix(countData = atac.pseudobulk[grepl("chrX", row.names(atac.pseudobulk)) ==F,],
                                     colData = atac.info,
                                     design = ~ timepoint)

atac.deseq <- DESeq(atac.deseq)

saveRDS(atac.deseq, "~/Mouse_Cereb/012_devPeaks/Mouse_atac_pseudobulk_byTimepoint_deseq.rds")
```

Let's have a look into some MA plots

```{r}
atac.deseq <- readRDS("~/Mouse_Cereb/012_devPeaks/Mouse_atac_pseudobulk_byTimepoint_deseq.rds")

plotMA(atac.deseq, contrast=c("timepoint","e10", "e11"))
plotMA(atac.deseq, contrast=c("timepoint","e11", "e12"))
plotMA(atac.deseq, contrast=c("timepoint","e12", "e13"))
plotMA(atac.deseq, contrast=c("timepoint","e13", "e15"))
plotMA(atac.deseq, contrast=c("timepoint","e15", "e17"))
plotMA(atac.deseq, contrast=c("timepoint","e17", "P0"))
plotMA(atac.deseq, contrast=c("timepoint","P0", "P04"))
plotMA(atac.deseq, contrast=c("timepoint","P04", "P07"))
plotMA(atac.deseq, contrast=c("timepoint","P07", "P14"))
plotMA(atac.deseq, contrast=c("timepoint","P14", "P63"))
```

```{r, eval=FALSE}
atac.deseq.res <- data.frame(peak=row.names(atac.deseq), stringsAsFactors = F)

getDeseqRes <- function(deseqObj, s1, s2, slot="timepoint") {
  res <- results(deseqObj, contrast=c("timepoint", s2, s1), alpha = 0.05, lfcThreshold = 0.5)
  atac.deseq.res[, paste(s1, s2, sep = "_")] <<- ifelse(res$padj < 0.05 & abs(res$log2FoldChange) > 0.5,
                     yes = ifelse(res$log2FoldChange > 0,
                                  yes = "up",
                                  no = "down"),
                     no="no")
  #atac.deseq.res <<- left_join(atac.deseq.res, select(res, peak, paste(s1, s2, sep = "-")))
  return(res)
}

atac.deseq.res_list <- lapply(1:(length(timepoints)-1), function(t) getDeseqRes(atac.deseq, timepoints[t], timepoints[t+1]))
atac.deseq.res[is.na(atac.deseq.res)] <- "not-tested"

saveRDS(atac.deseq.res_list, "~/Mouse_Cereb/012_devPeaks/Mouse_atac_pseudobulk_byTimepoint_deseq_results_list.rds")

saveRDS(atac.deseq.res, "~/Mouse_Cereb/012_devPeaks/Mouse_atac_pseudobulk_byTimepoint_deseq_results.rds")
```

```{r, fig.height=4, fig.width=6}
atac.deseq.res <- readRDS("~/Mouse_Cereb/012_devPeaks/Mouse_atac_pseudobulk_byTimepoint_deseq_results.rds")

atac.deseq.res.gather <- gather(atac.deseq.res, -peak, key = "timepoint", value = "sign", na.rm = F) %>%
  filter(sign %in% c("up", "down")) %>%
  group_by(timepoint, sign, .drop = FALSE) %>%
    dplyr::count() %>%
  ungroup()

atac.deseq.res.gather$timepoint <- factor(atac.deseq.res.gather$timepoint, levels = sapply(1:(length(timepoints)-1), function(t) paste(timepoints[t], timepoints[t+1], sep = "_")))

atac.deseq.res.gather$stage <- as.integer(atac.deseq.res.gather$timepoint)

p <- ggplot(atac.deseq.res.gather, aes(x=stage, y=n, color=sign)) +
  geom_point() +
  geom_line(lwd = 0.8) +
  scale_x_continuous(breaks = seq(min(atac.deseq.res.gather$stage), max(atac.deseq.res.gather$stage), by = 1), labels=levels(atac.deseq.res.gather$timepoint)) +
  scale_color_manual(values = c("indianred", "deepskyblue3")) +
  theme_classic() +
  ylab("Number of DA regions") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), panel.grid.minor.x = element_blank(), panel.grid.major.x = element_blank())

p

if(dir.exists("~/Mouse_Cereb/Figures/004_devPeaks")==F) {
  dir.create("~/Mouse_Cereb/Figures/004_devPeaks", recursive = T)
}

pdf("~/Mouse_Cereb/Figures/004_devPeaks/Mouse_peaks_deseq_byStage.pdf", width = 6, height = 4, useDingbats = F); print(p); dev.off()
```

### Differences in cellular composition

```{r}
cell_anno <- readRDS("~/Mouse_Cereb/004_cellTypes_broad/Mouse_Cerebellum_atac_finalCellType_annotation.rds")

## We will exclude the mixtures "Glut_DN+Isthmic_nuclei", "Parabrachial+Isthmic_nuclei" as well as the heterogeneous group "Other"

cell_comp <- filter(cell_anno, !Cell_type_broad %in% c("Parabrachial+Isthmic_nuclei", "Glut_DN+Isthmic_nuclei", "Other")) %>%
  group_by(Sample_short, Cell_type_broad) %>%
  dplyr::count() %>%
  spread(key = "Sample_short", value = "n", fill = 0) %>%
  column_to_rownames("Cell_type_broad") %>%
  as.matrix()

cell_comp.f <- t(t(cell_comp)/colSums(cell_comp))

## Estimating differences in cellular composition across samples
deltaF <- do.call(rbind, lapply(1:ncol(cell_comp.f), function(i) {
  sapply(1:ncol(cell_comp.f), function(j) {
      df <- sum(abs(cell_comp.f[,i] - cell_comp.f[,j]))
  return(df)
  })
}))

colnames(deltaF) <- colnames(cell_comp.f)
row.names(deltaF) <- colnames(cell_comp.f)

pheatmap::pheatmap(deltaF, cluster_rows = F, cluster_cols = F)

compareDeltaF <- function(t1, t2) {
  t.test(as.numeric(deltaF[grepl(paste0("^",t1, "_"), row.names(deltaF)), grepl(paste0("^",t2, "_"), row.names(deltaF))]),
       c(as.numeric(deltaF[grepl(paste0("^",t1, "_"), row.names(deltaF)), grepl(paste0("^",t1, "_"), row.names(deltaF))])[2],
         as.numeric(deltaF[grepl(paste0("^",t2, "_"), row.names(deltaF)), grepl(paste0("^",t2, "_"), row.names(deltaF))])[2]))
}

timepoints <- c("e10", "e11", "e12", "e13", "e15", "e17", "P0", "P04", "P07", "P14", "P63")

deltaF.stats <- Reduce(bind_rows, lapply(1:(length(timepoints)-1), function(i) {
  x <- compareDeltaF(timepoints[i], timepoints[i+1])
  return(data.frame(timepoint=paste(timepoints[i], timepoints[i+1], sep = "-"), diff_mean=x$estimate[1], diff_mean_rep=x$estimate[2], diff_ci_low=x$conf.int[1], diff_ci_up=x$conf.int[2], pval=x$p.value, t=x$statistic, stringsAsFactors = F))
}))

deltaF.stats$timepoint <- factor(deltaF.stats$timepoint)
deltaF.stats$stage <- as.numeric(deltaF.stats$timepoint)

p <- ggplot(deltaF.stats) +
  geom_point(aes(x=timepoint, y=diff_mean, size=-log10(pval)), color="deepskyblue3") +
  geom_point(aes(x=timepoint, y=diff_mean_rep), color="darkorange", position = position_jitter(width = 0.2, height = 0, seed = 3)) +
  ylab("Difference in cell type proportions") +
  xlab("Developmental stage") +
  geom_hline(yintercept = mean(deltaF.stats$diff_mean), color="lightskyblue", lty="dashed") +
    geom_hline(yintercept = mean(deltaF.stats$diff_mean_rep), color="orange", lty="dashed") +
  theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), panel.grid.minor.x = element_blank(), panel.grid.major.x = element_blank())

p

pdf("~/Mouse_Cereb/Figures/004_devPeaks/Mouse_peaks_deseq_byStage_CellComposition_DeltaFraction_Ttest.pdf", width = 6, height = 4, useDingbats = F); print(p); dev.off()
```


```{r}
sessionInfo()
```

