---
title: "Mouse Cerebellum - developmental differences in prenatal EGL"
author: "Ioannis Sarropoulos"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(Seurat)
  library("DESeq2")
  library("rGREAT")
  library(maSigPro)
  library(RColorBrewer)
  library(viridis)
  library(ComplexHeatmap)
  library(Mfuzz)
  library(UpSetR)
  library(Matrix)
})
```

```{r}
lineage <- "GC"
```

Here we are continuing our deconvolution of the additive effect between development and differentiation in GCs.

We have already seen that we can explain differences in later timepoints by running DESeq2. We find lots of changes in P0-P4, P7-P14 and P14-P63, which also show consistent differences in conservation (see DESeq2 analysis).

However, in our original analysis we also saw gradual differences in early stages (e13-P0). Here we didn't detect changes with DESeq2, likely because the degree of change between individual stages is small.

Thus, we want to look for smaller, but consistent changes across time in the same cell type/state.

For this, maSigPro and Mfuzz (the methods used in the evodevo project) are ideal to allow us to answer this question.

We will focus on a specific cell state, namely the early differentiation stage for e13-P0 (EGL).

This also controls for potential artifacts in the harmony integration, as here we will be considering unintegrated data!

```{r}
## Mouse peak info
mouse_peaks <- read.table("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peaks_summary.txt", header = T, sep = "\t", stringsAsFactors = F)

mouse_peaks$min_age_collapsed <- factor(mouse_peaks$min_age_collapsed, levels = c("0-20_muridae", "73-105_eutherian", "160-177_mammalian", "312-350_tetrapod", "435_vertebrate"))

## Pseudotime seurat and bins
atac.seurat <- readRDS(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_lineage_atac_seurat.rds"))

pseudotime_bins <- read_tsv(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_pseudotime_stats.txt"))

pseudotime_bins$timepoint <- factor(pseudotime_bins$timepoint, levels = c("e10", "e11", "e12", "e13", "e15", "e17", "P0", "P04", "P07", "P14", "P63"))

## Differentiation cluster (Mfuzz in pseudotime)
peaks_clust <- readRDS(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/Mouse_within", lineage, "_lineage_peaks_Mfuzz_cluster_object.rds"))

peak_stats <- data.frame(peak=names(peaks_clust$cluster), mfuzz=peaks_clust$cluster, stringsAsFactors = F )

## Function to plot genes
plotMarkers <- function(gene, obj=atac.seurat) {
  ggplot(NULL, aes(x=obj@reductions$umap@cell.embeddings[,1],
                 y=obj@reductions$umap@cell.embeddings[,2],
                 col=log10(obj@assays$GeneScoreScaled@data[gene,]))) +
  geom_point(size=0.1) +
  scale_colour_viridis_c(na.value = "gray90", option = "magma", direction = 1, name="log10 CPM") +
  xlab("UMAP1") +
  ylab("UMAP2") +
  ggtitle(paste0("Promoter_", gene))+
  theme_classic()
}

if(dir.exists(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/"))==F) {
  dir.create(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/"))
}
```

Exploring the Seurat object

```{r}
DimPlot(atac.seurat, group.by = "clusters_original", label = T)
DimPlot(atac.seurat, group.by = "timepoint", label = T)

plotMarkers("Atoh1")
plotMarkers("Gli2")
plotMarkers("Barhl1")
```

We can focus our analysis on the original cluster 4 (early dev. stages EGL).

```{r}
table(atac.seurat$timepoint[atac.seurat$clusters_original=="4"])
```


We should focus on e13-P0, we only have very few cells from the other timepoints.

Let's generate pseudobulks by sample (timepoint and replicate within this cluster).
As for other analyses, we will only consider robust peaks here:

```{r}
atac.info <- atac.seurat@meta.data %>%
  rownames_to_column("cell") %>%
  mutate(sample=paste(timepoint, orig.ident, sep = "_")) %>%
  arrange(sample) %>%
  filter(clusters_original=="4", timepoint %in% c("e13", "e15", "e17", "P0"))

group_by(atac.info, sample) %>%
  dplyr::count()


ggplot(atac.info, aes(x=timepoint, y=pseudotime, fill=timepoint)) +
  geom_boxplot(notch = T) +
  scale_fill_manual(values = brewer.pal(name = "Spectral", n=length(unique(atac.info$timepoint))), guide=F)+
  ylab("Pseudotime") +
  xlab("Pseudotime bins") +
  theme_classic()
```

We should have enough cells for each sample to perform this analysis.

Now let's generate pseudobulks

```{r}
atac.counts <- do.call(cbind,lapply(unique(atac.info$sample), function(x) {
  Matrix::rowSums(atac.seurat@assays$Peaks@counts[, atac.info$cell[atac.info$sample==x]])
}))

dim(atac.counts)

colnames(atac.counts) <- unique(atac.info$sample)

hist(log10(apply(atac.counts, 1, max) + 1), breaks=30)

table(rowSums(atac.counts >= 5))
table(rowSums(atac.counts >= 10))
```

Here we are looking for small quantitative differences so we have to go for relatively robust expression.

Let's only keep peaks that reach at least 10 counts in at least two samples (i.e. reps)

```{r}
atac.counts.robust <- atac.counts[rowSums(atac.counts >= 10) >=2,]

dim(atac.counts.robust)

sum(row.names(atac.counts.robust) %in% gsub("_", "-", mouse_peaks$peak[mouse_peaks$robust]))

## Most of them are already scored as robust. Let's also subset for that just to be sure.
atac.counts.robust <- atac.counts.robust[row.names(atac.counts.robust) %in% gsub("_", "-", mouse_peaks$peak[mouse_peaks$robust]),]

dim(atac.counts.robust)

## Scaling peaks using VST from DESeq2
atac.vst <- varianceStabilizingTransformation(atac.counts.robust)
```

# Identifying temporally dynamic peaks in EGL with maSigPro

Let's make a design matrix for maSigPro. 

For this, we also have to specify a time component.
Conveniently, here we are looking at regular intervals of two days (e13-e15-e17-P0) so we can simply specify time as integers 1:4.

We also declare all samples as members of the same group, as we are only interested in differences across time in an otherwise homogeneous cell population.

We can run the model directly on the VST normalised data, using a negative binomial distribution for the GLM model:

```{r, eval=FALSE}
design_mat <- data.frame(Time=rep(1:4, each=2), Replicate=rep(1:4, each=2), Group=1, row.names = colnames(atac.vst), stringsAsFactors = F)

design_mat

## We are using second degree polynomials since we only have four timepoints
d <- make.design.matrix(design_mat, degree = 2, time.col = 1, repl.col = 2, group.cols = 3)

## Fitting the model
set.seed(1)
NBp <- p.vector(atac.vst, d, counts=TRUE, Q = 0.05, MT.adjust = "BH", min.obs = 2)

NBt <- T.fit(NBp)

## Getting data for all genes, then we can selecte a meaningful Rsqrd/pval threshold.
get_all<-get.siggenes(NBt, rsq = 0, vars="all")

head(get_all$sig.genes$sig.pvalue)

hist(get_all$sig.genes$sig.pvalue$`p-value`)
hist(get_all$sig.genes$sig.pvalue$`R-squared`)

sum(get_all$sig.genes$sig.pvalue$`R-squared` > 0.3)
sum(get_all$sig.genes$sig.pvalue$`R-squared` > 0.7)

saveRDS(NBp, "~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_NBp.rds")
saveRDS(NBt, "~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_NBt.rds")
saveRDS(get_all, "~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_allPeaks.rds")

# Let's be conservative and go for peaks with Rsqrd > 0.7.
sig.peaks <- row.names(get_all$sig.genes$sig.pvalues)[get_all$sig.genes$sig.pvalue$`R-squared` > 0.7]
length(sig.peaks)

write(sig.peaks, "~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_significantPeaks_Rsq0.7.txt")
```

```{r}
get_all <- readRDS("~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_allPeaks.rds")
sig.peaks <- readLines("~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_significantPeaks_Rsq0.7.txt")
```

Now we need to cluster the most significant peaks based on their temporal profile. For that we need to calculate average values per timepoint.

```{r}
atac.vst_byTimepoint <- atac.vst[sig.peaks,] %>%
  as.data.frame() %>%
  rownames_to_column("peak") %>%
  gather(-peak, key = "sample", value = "vst") %>%
  separate(sample, into = c("timepoint", "rep")) %>%
  group_by(timepoint, peak) %>%
  summarise(vst=mean(vst)) %>%
  spread(key = timepoint, value = vst) %>%
  column_to_rownames("peak") %>%
  as.matrix()
```

Now we can use Mfuzz to cluster these peaks. We will use 2 clusters

```{r, fig.width=5, fig.height=5, eval=FALSE}
## Constructing an eset object
atac.vst_byTimepoint <- ExpressionSet(atac.vst_byTimepoint)

## Standardising (essentially substracting the mean and dividing by standard deviation).
atac.vst_byTimepoint.s <- standardise(atac.vst_byTimepoint)

m1 <- mestimate(atac.vst_byTimepoint.s)
m1

set.seed(1)
c_dmin <- Dmin(atac.vst_byTimepoint.s,m1,crange=seq(2,10,1),repeats = 5)
plot(c_dmin)

set.seed(1)
temp_clust <- mfuzz(atac.vst_byTimepoint.s,c=2,m=m1)

mfuzz.plot2(atac.vst_byTimepoint.s, temp_clust, colo = "fancy", x11 = F, centre = T, mfrow = c(2,2))
table(temp_clust$cluster)

saveRDS(atac.vst_byTimepoint.s, "~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_significantPeaks_Rsq0.7_VST_byTimepoint_standardised.rds")

saveRDS(temp_clust, "~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_significantPeaks_Rsq0.7_Mfuzz_clusters_k2.rds")
```

```{r}
atac.vst_byTimepoint.s <- readRDS("~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_significantPeaks_Rsq0.7_VST_byTimepoint_standardised.rds")

temp_clust <- readRDS("~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/Mouse_GC_EGL_e13-P0_masigPro_significantPeaks_Rsq0.7_Mfuzz_clusters_k2.rds")

mfuzz.plot2(atac.vst_byTimepoint.s, temp_clust, colo = "fancy", x11 = F, centre = T, mfrow = c(1,2))
table(temp_clust$cluster)
```

```{r}
peak_stats <- data.frame(peak=names(temp_clust$cluster), temp_clust=temp_clust$cluster, stringsAsFactors = F) %>%
  full_join(peak_stats)

peak_stats <- mutate(mouse_peaks, peak=gsub("_", "-", peak)) %>%
  right_join(peak_stats)
```

Now let's look how these peaks look like in terms of conservation:

# Sequence constraint and age of temporally dynamic peaks

Ordering clusters from early to late:

```{r}
clust_order <- data.frame(clust=1:nrow(temp_clust$centers),
                          max_activity=apply(temp_clust$centers, 1, which.max),
                          inflex=apply(temp_clust$centers, 1, function(x) which.min(abs(x)))) %>%
  arrange(max_activity, inflex)


clust_order
```

```{r}
peak_stats$temp_clust <- factor(peak_stats$temp_clust, levels=clust_order$clust)

ggplot(filter(peak_stats,genomic_class_broad %in% c("Distal"), is.na(temp_clust)==F), aes(x=temp_clust, y=meanPhastCons_100bp, fill=temp_clust)) +
  geom_boxplot(notch = T) +
  scale_fill_manual(values = c("indianred","deepskyblue3"), name="Temporal cluster") +
  scale_x_discrete(labels=c("down", "up")) +
  ylab("PhastCons (distal CREs)") +
  xlab("DA direction") +
  theme_classic()

ggplot(filter(peak_stats,genomic_class_broad %in% c("Distal"), is.na(temp_clust)==F), aes(temp_clust,  fill=as.factor(min_age_num)))+
  geom_bar(position = "fill") +
  scale_x_discrete(labels=c("down", "up")) +
  ylab("Fraction") +
  xlab("DA direction") +
  scale_fill_manual(values=colorRampPalette(brewer.pal(11, "Spectral"))(length(unique(peak_stats$min_age_num)))) +
  theme_classic()
```

Now let's look into these peaks for GREAT and Homer enrichments:

# GO enrichment of associated (proximal) genes

GREAT

```{r, eval=FALSE}
makeBed <- function(peaks) {
  df <- as.data.frame(do.call(rbind, lapply(peaks, function(p) {
    return(c(str_split(p, pattern = "-", simplify = T), p))
  })), stringsAsFactors=F)
  df[,2] <- as.numeric(df[,2])
  df[,3] <- as.numeric(df[,3])
  colnames(df) <- c("chr", "start", "end", "peak")
  return(df)
}

bg_all <- makeBed(names(temp_clust$cluster))

allPeaks_enrichments <- lapply(unique(temp_clust$cluster), function(clust) {
  peaks <- makeBed(names(temp_clust$cluster[temp_clust$cluster==clust]))
  print(paste0("Submitting job for cluster ", clust))
  job <- submitGreatJob(peaks, bg = bg_all, species = "mm10", includeCuratedRegDoms = F, request_interval = 10)
  tb <- getEnrichmentTables(job, availableOntologies(job))
  print(paste0("Getting enrichment for cluster ", clust))
  saveRDS(tb, paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/Mouse_peaks_E13-P0_dynamic_inEGL_", lineage, "_cluster_", clust, "_GREAT_all.rds"))
  print(paste0("Done with cluster ", clust))
  return(tb)
})
```

```{r, fig.width=6, fig.height=5}
great_bp <- Reduce(bind_rows,lapply(unique(temp_clust$cluster), function(clust) {
  tb <- readRDS(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/Mouse_peaks_E13-P0_dynamic_inEGL_", lineage, "_cluster_", clust, "_GREAT_all.rds"))
  bp <- tb[["GO Biological Process"]]
  bp$temp_cluster <- clust
  return(bp)
}))

great_bp <- arrange(great_bp, temp_cluster, Hyper_Adjp_BH)

length(unique(great_bp$ID))

great_bp_top <- group_by(great_bp, temp_cluster) %>%
  top_n(10, -Hyper_Adjp_BH)

length(unique(great_bp_top$ID))

dplyr::select(great_bp_top, ID, name, Hyper_Adjp_BH, temp_cluster)

great_bp_plot <- filter(great_bp, ID %in% great_bp_top$ID)

great_bp_plot_spread <- dplyr::select(great_bp_plot, ID, name, Hyper_Adjp_BH, temp_cluster) %>%
  spread(key = temp_cluster, value = Hyper_Adjp_BH)

great_bp_plot_spread[,3:ncol(great_bp_plot_spread)] <- -log10(great_bp_plot_spread[,3:ncol(great_bp_plot_spread)])

## Capping at 10
great_bp_plot_spread[,3:ncol(great_bp_plot_spread)][great_bp_plot_spread[,3:ncol(great_bp_plot_spread)] > 10] <- 10

row.names(great_bp_plot_spread) <- str_trunc(great_bp_plot_spread$name, width = 60)

pheatmap::pheatmap(great_bp_plot_spread[,3:ncol(great_bp_plot_spread)][clust_order$clust], cluster_rows = T, cluster_cols = F, color = viridis(100,option = "D"))

```

## TF activity

HOMER for TF motif enrichment

```{r, eval=FALSE}
makeBed <- function(peaks) {
  df <- as.data.frame(do.call(rbind, lapply(peaks, function(p) {
    return(c(str_split(p, pattern = "-", simplify = T), p))
  })), stringsAsFactors=F)
  df[,2] <- as.numeric(df[,2])
  df[,3] <- as.numeric(df[,3])
  colnames(df) <- c("chr", "start", "end", "peak")
  return(df)
}

peak_stats.only <- peak_stats[is.na(peak_stats$temp_clust)==F,]

if(dir.exists(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/homer/"))==F){
  dir.create(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/homer/"))
}

lapply(unique(peak_stats.only$temp_clust), function(clust) {
  peaks <- makeBed(peak_stats.only$peak[peak_stats.only$temp_clust==clust])
  write.table(peaks,paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/homer/Mouse_peaks_E13-P0_dynamic_inEGL_", lineage, "_cluster_", clust,"_all.bed"), col.names = F, row.names = F, sep = "\t", quote = F)
})

write.table(makeBed(peak_stats.only$peak), paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/homer/Mouse_peaks_E13-P0_dynamic_inEGL_", lineage, "_all_background.bed"), col.names = F, row.names = F, sep = "\t", quote = F)
```

Running homer
```{bash, eval=FALSE}
#sh ~/Mouse_Cereb/015_withinLineages/GC/masigpro_stages/homer/Mouse_peaks_Homer_motifs_all.sh&
```

Now loading the results:

```{r}
motif_path <- list.files(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/homer/all/"), pattern = "Mouse_.*_all_Homer")

print(motif_path)

motifs <- Reduce(bind_rows,lapply(motif_path, function(p) {
  motifs <- read_tsv(paste0("~/Mouse_Cereb/015_withinLineages/", lineage, "/masigpro_stages/homer/all/", p, "/knownResults.txt"))
  motifs$lin_clust <- str_split(p, '_', simplify = T)[8]
  return(motifs)
}))

motifs <-   mutate(motifs, `Motif Name`= gsub("\\/.*", "", `Motif Name`)) %>%
  group_by(`Motif Name`, lin_clust) %>%
  summarise(`Log P-value`=min(`Log P-value`)) %>%
  ungroup()

## All motifs included in the file are significant!

motifs_spread <- dplyr::select(motifs, `Motif Name`, `Log P-value`, lin_clust) %>%
  spread(key = lin_clust, value=`Log P-value`)

## Now also identifying the top motif per cluster. We want to label them on the heatmap
top_motif_by_clust <- unique(apply(motifs_spread[,2:ncol(motifs_spread)], 2, function(x) motifs_spread$`Motif Name`[which.min(x)]))

top_motif_by_clust

## These aren't that informative
## Let's make our custom list

interesting_motifs <- c("Atoh1(bHLH)", "NF1(CTF)", "CTCF(Zf)",  "PAX6(Homeobox)", "Meis1(Homeobox)", "Sox2(HMG)", "ZEB1(Zf)",   "Gli2(Zf)", "RFX(HTH)", "Ets1-distal(ETS)", "STAT4(Stat)", "Brn1(POU,Homeobox)")

## Capping to 50
motifs_spread[,2:ncol(motifs_spread)] <- motifs_spread[,2:ncol(motifs_spread)] * (-1)
motifs_spread[,2:ncol(motifs_spread)][motifs_spread[,2:ncol(motifs_spread)] > 50] <- 50

sum(apply(motifs_spread[,2:ncol(motifs_spread)], 1, max) > 10)

## Also limiting the analysis in highly significant motifs
motifs_spread.sig <- motifs_spread[apply(motifs_spread[,2:ncol(motifs_spread)], 1, max) > 10,]

## Annotating with the TF family
motif_anno <- data.frame(motif=motifs_spread.sig$`Motif Name`, stringsAsFactors = F) %>%
  mutate(tf_family=gsub("(", "", str_extract(motif, "\\(.*\\)"), fixed = T)) %>%
   mutate(tf_family=gsub(")", "", tf_family, fixed = T))

row.names(motif_anno) <- motif_anno$motif

## Keeping only prevalent families and setting the rest to other
top_families <- group_by(motif_anno, tf_family) %>%
  dplyr::count() %>%
  arrange(desc(n))

top_families <- top_families$tf_family[1:10]

motif_anno$tf_family[!(motif_anno$tf_family %in% top_families)] <- "Other"

n <- brewer.pal(11, "Set3")
names(n) <- unique(motif_anno$tf_family)
```

```{r, fig.width=6, fig.height=6}
ha = rowAnnotation(foo = anno_mark(at = which(motifs_spread.sig$`Motif Name` %in% interesting_motifs), labels = motifs_spread.sig$`Motif Name`[which(motifs_spread.sig$`Motif Name` %in% interesting_motifs)]), annotation_legend_param = list(direction = "vertical",nrow = 1, labels_gp = gpar(fontsize = 0.1), padding = unit(10, "mm")))

draw(Heatmap(matrix = as.matrix(motifs_spread.sig[,2:ncol(motifs_spread)][,clust_order$clust]),
        cluster_rows = T,
        cluster_columns = F,
        show_column_names = T,
        clustering_distance_columns = "pearson",
        clustering_distance_rows = "pearson",
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2",
        show_row_names = F,
        col=viridis(100,option = "E"),
        border = TRUE,
        left_annotation = rowAnnotation(tf_family=motif_anno$tf_family, col=list(tf_family=n), border=T, annotation_legend_param = list(direction = "vertical")),
        right_annotation = ha,
        heatmap_legend_param = list( title = "-log10 P-val", direction="vertical")), heatmap_legend_side = "right")

```

Now we see more clearly the difference between the homeobox/HMG bound peaks and the bHLH/Zf EGL-specific peaks!

Now we can also load the merged dataset and see where these peaks are active across other cells.

```{r}
peak_mat <- readRDS("~/Mouse_Cereb/006_peaks/Mouse_Cerebellum_reproducible_peak_matrix_SE_withInfo_robustPeaks.rds")
atac.data <- assay(peak_mat)
row.names(atac.data) <- gsub("_", "-", peak_mat@elementMetadata$peak)

umap <- readRDS("~/Mouse_Cereb/003_finalNucSet/Mouse_Cereb_finalNucSet_UMAP_merged_tiles.rds")
umap <- umap$df
colnames(umap) <- c("UMAP1", "UMAP2")
umap$cell <- row.names(umap)
```

Per cell in the embedding, we will calculate the fraction of reads in peaks of each cluster (using all peaks as background to allow for the effect of peaks not present in GC cells). The we can project the scores on the embedding.

```{r}
temp_clust_fraction <- Reduce(bind_rows, lapply(unique(temp_clust$cluster), function(c) {
  peaks <- names(temp_clust$cluster)[temp_clust$cluster==c]
  f <- Matrix::colSums(atac.data[peaks,])/Matrix::colSums(atac.data)
  ## Putting in a dataframe and return
  df <- data.frame(cell=colnames(atac.data), fraction=f, peak_cluster=c, stringsAsFactors = F)
  return(df)
}))

umap <- left_join(umap, temp_clust_fraction)
umap$peak_cluster <- factor(umap$peak_cluster, levels = clust_order$clust)

ggplot(umap, aes(x=UMAP1, y=UMAP2, color=fraction)) +
  geom_point(size=0.1, alpha=0.3) +
  scale_color_viridis() +
  facet_wrap(~peak_cluster) +
  theme_classic()
```

We clearly see that the elements that go down (more conserved) are pleiotropic elements that were previously active in the progenitors (e10-e12) and even appear active in Purkinje and interneuron neuroblasts.

Instead, elements going up with time (less conserved) are specific to GCs (and mostly limited to the EGL, thus not confounded by the differentiation signal).

Finally, we have to see whether these peaks that show differences across timepoints are also associated with gene expression changes (i.e. do they have functional consequences or are they just remnants of what the cell was doing before or what it is about to do next)?

Since the question is whether genes show a pattern that matches the peak activity, we can't use correlation approaches to assign peaks to genes. Instead, it makes mostly sense to look for the activity of proximal genes.

#Activity of putative target genes

We will use assignment to the closest TSS.

```{r}
## Identifying peaks dynamic across pseudotime and with a temporal pattern between e13-P0:
peaks_by_gene <- filter(peak_stats, is.na(temp_clust)==F, is.na(mfuzz)==F) %>%
  dplyr::select(peak, temp_clust, associated_gene) %>%
  group_by(temp_clust, associated_gene) %>%
  dplyr::count()

length(peaks_by_gene$n)
## Total genes associated with a peak
length(unique(peaks_by_gene$associated_gene))

## Number of unique genes per cluster
group_by(peaks_by_gene, temp_clust) %>%
  summarise(x=n_distinct(associated_gene))

closest_genes_list <- lapply(unique(peaks_by_gene$temp_clust), function(x) {
  peaks_by_gene$associated_gene[peaks_by_gene$temp_clust==x]
})

names(closest_genes_list) <- unique(peaks_by_gene$temp_clust)

upset(fromList(closest_genes_list),nsets = 6,
      keep.order = T)
```

This seems like a reasonable number.

Now we can remove genes associated with peaks from more than one clusters.

```{r}
clust_by_gene <- group_by(peaks_by_gene, associated_gene) %>%
  summarise(nClust=n_distinct(temp_clust))

table(clust_by_gene$nClust)

peaks_by_gene_unique <- filter(peaks_by_gene, associated_gene %in% clust_by_gene$associated_gene[clust_by_gene$nClust==1])

## Number of genes to be used for this analysis:
nrow(peaks_by_gene_unique)
```

Now we can look at the activity of these genes in the data from Vladoiu et al.

We don't have any pseudotime data here, but we can look within clusters of (in theory) similar cell states to identify time-dependent signals.

```{r}
vladoiu_anno <- read_tsv("~/Mouse_Cereb/011_integrate_RNAseq/Vladoiu_2019/Vladoiu_cell_type_annotation.txt")

table(vladoiu_anno$Vladoiu_cell_type)

vladoiu_gc <- filter(vladoiu_anno, Vladoiu_cell_type %in% c("Embryonic and postnatal GCPs-1",
                                                       "Granule cells",
                                                       "Postnatal GCPs-2",
                                                       "Unipolar brush cell and GCP progenitor"
                                                       ))

ggplot(vladoiu_gc, aes(Vladoiu_cell_type, fill=timepoint)) +
  geom_bar() + 
  coord_flip() +
  theme_classic() +
  scale_fill_manual(values = brewer.pal(name = "Spectral", n=11)[c(4:7, 9)])
```

We will specifically look into the embryonic and postnatal GCPs and UBC/GCP progenitors (these are the ones where we see mixing of timepoints).

We only need to look into E14, E16, E18 and P0:

```{r}
vladoiu_clust_fraction <- Reduce(bind_rows,lapply(unique(peaks_by_gene_unique$temp_clust), function(c) {
  c_fraction <- Reduce(bind_rows, lapply(c("GSM3318001_E14", "GSM3318002_E16", "GSM3318003_E18", "GSM3318004_P0"), function(t) {
  ## Reading in the data
  rna.data <- readMM(paste0("~/Mouse_Cereb/011_integrate_RNAseq/Vladoiu_2019/Vladoiu_data/",t,"_matrix.mtx"))
  colnames(rna.data) <- readLines(paste0("~/Mouse_Cereb/011_integrate_RNAseq/Vladoiu_2019/Vladoiu_data/",t,"_barcodes.tsv"))
  ## Adding the timepoint info to the barcodes
  vladoiu_t <- str_split(t, "_", simplify = T)[2]
  colnames(rna.data) <- paste(vladoiu_t, colnames(rna.data), sep = "_")
## Also stripping the "-1" ending
colnames(rna.data) <- gsub("-1", "",colnames(rna.data))
  rna.genes <- read_tsv(paste0("~/Mouse_Cereb/011_integrate_RNAseq/Vladoiu_2019/Vladoiu_data/",t,"_genes.tsv"), col_names = F)
  row.names(rna.data) <- rna.genes$X2
  
  ## Subsetting for the cells in interest
  rna.data <- rna.data[, vladoiu_anno$cell[vladoiu_anno$Vladoiu_cell_type %in% c("Embryonic and postnatal GCPs-1","Unipolar brush cell and GCP progenitor") & vladoiu_anno$timepoint == vladoiu_t ]]
  
  ## Calculating fraction of reads in target genes of cluster 4
  comm.genes <- intersect(unique(peaks_by_gene_unique$associated_gene[peaks_by_gene_unique$temp_clust==c]), row.names(rna.data))
  f <- Matrix::colSums(rna.data[comm.genes,])/Matrix::colSums(rna.data)
  ## Putting in a dataframe and return
  df <- data.frame(cell=colnames(rna.data), f, stringsAsFactors = F)
  return(df)
}))
  c_fraction$temp_cluster <- c
  return(c_fraction)
}))

vladoiu_clust_fraction <- left_join(vladoiu_clust_fraction, vladoiu_anno)
```

```{r}
ggplot(filter(vladoiu_clust_fraction, Vladoiu_cell_type=="Embryonic and postnatal GCPs-1"), aes(x=timepoint, y=f, fill=timepoint)) +
  geom_boxplot(notch = T)+
  theme_classic() +
  ylab("Fraction of UMIs in target genes of cluster") +
  scale_fill_manual(values = brewer.pal(name = "Spectral", n=4)) +
  facet_wrap(~as.factor(temp_cluster), scales = "free_y", nrow = 1)+
  theme(legend.position = "top")

ggplot(filter(vladoiu_clust_fraction, Vladoiu_cell_type=="Unipolar brush cell and GCP progenitor"), aes(x=timepoint, y=f, fill=timepoint)) +
  geom_boxplot(notch = T)+
  theme_classic() +
  ylab("Fraction of UMIs in target genes of cluster") +
  scale_fill_manual(values = brewer.pal(name = "Spectral", n=4)) +
  facet_wrap(~as.factor(temp_cluster), scales = "free_y", nrow = 1)+
  theme(legend.position = "top")
```

```{r}
## Up
wilcox.test(vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==2 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E14"], vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==2 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E16"])

wilcox.test(vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==2 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E16"], vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==2 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E18"])

wilcox.test(vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==2 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E18"], vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==2 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="P0"])

## Down
wilcox.test(vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==1 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E14"], vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==1 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E16"])

wilcox.test(vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==1 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E16"], vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==1 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E18"])

wilcox.test(vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==1 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="E18"], vladoiu_clust_fraction$f[vladoiu_clust_fraction$temp_cluster==1 & vladoiu_clust_fraction$Vladoiu_cell_type=="Embryonic and postnatal GCPs-1" & vladoiu_clust_fraction$timepoint=="P0"])
```


We see that the target genes follow the same activity pattern (especially those going down).

Thus, temporal differences in peak activity have an effect in the expression of their putative target genes!

```{r}
sessionInfo()
```

